// @remax/toutiao/cjs/hostComponents/Button/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.props = exports.alias = void 0;
exports.alias = {
    className: 'class',
    id: 'id',
    style: 'style',
    type: 'type',
    plain: 'plain',
    disabled: 'disabled',
    loading: 'loading',
    formType: 'form-type',
    openType: 'open-type',
    hoverClass: 'hover-class',
    hoverClassName: 'hover-class',
    hoverStopPropagation: 'hover-stop-propagation',
    hoverStartTime: 'hover-start-time',
    hoverStayTime: 'hover-stay-time',
    lang: 'lang',
    sessionFrom: 'session-from',
    sendMessageTitle: 'send-message-title',
    sendMessagePath: 'send-message-path',
    sendMessageImg: 'send-message-img',
    appParameter: 'app-parameter',
    showMessageCard: 'show-message-card',
    onGetUserinfo: 'bindgetuserinfo',
    onGetUserInfo: 'bindgetuserinfo',
    onContact: 'bindcontact',
    onGetPhoneNumber: 'bindgetphonenumber',
    onError: 'binderror',
    onOpenSetting: 'bindopensetting',
    onLaunchApp: 'bindlaunchapp',
    onTap: 'bindtap',
    onClick: 'bindtap',
    animation: 'animation',
};
exports.props = Object.values(exports.alias);
