// @remax/one/esm/hostComponents/Button/index.web.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import * as React from 'react';
import clsx from 'clsx';
import { filterProps } from '../../utils/isPlatformSpecifyProp';
import useWebTouch from '../useWebTouch';
var Button = function (props, ref) {
    var _a;
    var _b = filterProps(props), hoverClassName = _b.hoverClassName, _c = _b.hoverStartTime, hoverStartTime = _c === void 0 ? 400 : _c, _d = _b.hoverStayTime, hoverStayTime = _d === void 0 ? 50 : _d, className = _b.className, onTouchStart = _b.onTouchStart, onTouchMove = _b.onTouchMove, onTouchEnd = _b.onTouchEnd, onTouchCancel = _b.onTouchCancel, onTap = _b.onTap, onLongTap = _b.onLongTap, restProps = __rest(_b, ["hoverClassName", "hoverStartTime", "hoverStayTime", "className", "onTouchStart", "onTouchMove", "onTouchEnd", "onTouchCancel", "onTap", "onLongTap"]);
    var _e = __read(useWebTouch({
        hoverDelay: hoverStartTime,
        hoverDuration: hoverStayTime,
        onLongTap: onLongTap,
        onTouchStart: onTouchStart,
        onTouchMove: onTouchMove,
        onTouchEnd: onTouchEnd,
        onTouchCancel: onTouchCancel,
    }), 5), hovered = _e[0], handleTouchStart = _e[1], handleTouchMove = _e[2], handleTouchEnd = _e[3], handleTouchCancel = _e[4];
    return (React.createElement("button", __assign({}, restProps, { onTouchStart: handleTouchStart, onTouchMove: handleTouchMove, onTouchEnd: handleTouchEnd, ref: ref, onTouchCancel: handleTouchCancel, className: clsx('remax-button', className, (_a = {}, _a[hoverClassName || ''] = hovered, _a)), onClick: onTap })));
};
export default React.forwardRef(Button);
