// @remax/runtime/cjs/hooks/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useQuery = exports.useNativeEffect = exports.useComponentInstance = exports.usePageInstance = exports.useAppEvent = exports.usePageEvent = void 0;
var framework_shared_1 = require("@remax/framework-shared");
Object.defineProperty(exports, "usePageEvent", { enumerable: true, get: function () { return framework_shared_1.usePageEvent; } });
Object.defineProperty(exports, "useAppEvent", { enumerable: true, get: function () { return framework_shared_1.useAppEvent; } });
Object.defineProperty(exports, "usePageInstance", { enumerable: true, get: function () { return framework_shared_1.usePageInstance; } });
Object.defineProperty(exports, "useComponentInstance", { enumerable: true, get: function () { return framework_shared_1.useComponentInstance; } });
var useNativeEffect_1 = require("./useNativeEffect");
Object.defineProperty(exports, "useNativeEffect", { enumerable: true, get: function () { return __importDefault(useNativeEffect_1).default; } });
var useQuery_1 = require("./useQuery");
Object.defineProperty(exports, "useQuery", { enumerable: true, get: function () { return __importDefault(useQuery_1).default; } });
