// @remax/ali/esm/node/index.js
import * as path from 'path';
import hostComponents from '../hostComponents/node';
var TPL_DEFAULT_ROOT = path.join(__dirname, '../../templates', 'default');
var TPL_STATIC_ROOT = path.join(__dirname, '../../templates', 'static');
var plugin = function () {
    return {
        meta: {
            global: 'my',
            template: {
                extension: '.axml',
                tag: 'import',
                src: 'src',
            },
            style: '.acss',
            jsHelper: {
                extension: '.sjs',
                tag: 'import-sjs',
                src: 'from',
            },
            ejs: {
                base: '',
                page: path.join(TPL_DEFAULT_ROOT, 'page.ejs'),
            },
            staticEjs: {
                base: '',
                page: path.join(TPL_STATIC_ROOT, 'page.ejs'),
                isolatedTemplates: path.join(TPL_STATIC_ROOT, 'isolated-templates.ejs'),
            },
        },
        hostComponents: hostComponents,
        skipHostComponents: ['swiper-item', 'picker-view-column'],
    };
};
export default plugin;
