// @remax/web/cjs/bootstrapMpa.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var react_dom_1 = require("react-dom");
var umi_hd_1 = __importDefault(require("umi-hd"));
var createAppConfig_1 = __importDefault(require("./createAppConfig"));
var framework_shared_1 = require("@remax/framework-shared");
var createPageConfig_1 = __importDefault(require("./createPageConfig"));
var noop = function () {
    // ignore
};
function createApp(options) {
    var appComponent = options.appComponent, pageComponent = options.pageComponent, page = options.page, appConfig = options.appConfig;
    var AppConfig = createAppConfig_1.default(appComponent);
    var PageConfig = createPageConfig_1.default(pageComponent, page.route);
    return (React.createElement(AppConfig, null,
        React.createElement(PageConfig, { pageConfig: __assign(__assign({}, appConfig.window), page.config), tabBar: appConfig.tabBar, location: window.location, cacheLifecycles: {
                didCache: noop,
                didRecover: noop,
            } })));
}
function bootstrap(options) {
    umi_hd_1.default();
    var _a = options.plugins, plugins = _a === void 0 ? [] : _a, page = options.page;
    var pluginDriver = new framework_shared_1.PluginDriver(plugins.map(function (plugin) { return plugin.default || plugin; }));
    var history = {
        push: function (url) {
            url = history.transformUrl(url);
            console.log('not implemented');
        },
        go: function () {
            console.log('not implemented');
        },
        replace: function (url) {
            url = history.transformUrl(url);
            location.replace(url);
        },
        page: null,
        transformUrl: function (url) {
            if (!url)
                return url;
            var publicPath = '';
            if (url.startsWith('/'))
                publicPath = location.pathname.replace("/" + page.route + ".html", '');
            var arr = url.split('?');
            arr[0] = arr[0] + '.html';
            return publicPath + arr.join('?');
        },
    };
    framework_shared_1.RuntimeOptions.apply({ pluginDriver: pluginDriver, history: history, mpa: true });
    var App = createApp(options);
    react_dom_1.render(App, document.getElementById('remax-app'));
}
exports.default = bootstrap;
