// @remax/web/cjs/createAppConfig.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var react_is_1 = require("react-is");
var framework_shared_1 = require("@remax/framework-shared");
var DefaultAppComponent = /** @class */ (function (_super) {
    __extends(DefaultAppComponent, _super);
    function DefaultAppComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefaultAppComponent.prototype.render = function () {
        return React.createElement(React.Fragment, null, this.props.children);
    };
    return DefaultAppComponent;
}(React.Component));
function createAppConfig(App) {
    var createConfig = function (AppComponent) {
        if (AppComponent === void 0) { AppComponent = DefaultAppComponent; }
        var config = {
            _instance: React.createRef(),
            callLifecycle: function (lifecycle) {
                var _a;
                var args = [];
                for (var _i = 1; _i < arguments.length; _i++) {
                    args[_i - 1] = arguments[_i];
                }
                var callbacks = framework_shared_1.AppInstanceContext.lifecycleCallback[lifecycle] || [];
                var result;
                callbacks.forEach(function (callback) {
                    result = callback.apply(void 0, __spread(args));
                });
                if (result) {
                    return result;
                }
                var callback = framework_shared_1.callbackName(lifecycle);
                if (this._instance.current && this._instance.current[callback]) {
                    return (_a = this._instance.current)[callback].apply(_a, __spread(args));
                }
            },
        };
        return /** @class */ (function (_super) {
            __extends(AppConfig, _super);
            function AppConfig() {
                var _this = _super !== null && _super.apply(this, arguments) || this;
                _this.visibilityChangeEvent = function () {
                    if (document.visibilityState === 'visible') {
                        config.callLifecycle(framework_shared_1.AppLifecycle.show);
                    }
                    else {
                        config.callLifecycle(framework_shared_1.AppLifecycle.hide);
                    }
                };
                _this.handelVisibilityChange = function () {
                    document.addEventListener('visibilitychange', _this.visibilityChangeEvent);
                };
                return _this;
            }
            AppConfig.prototype.componentDidMount = function () {
                window.onerror = function () {
                    var params = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        params[_i] = arguments[_i];
                    }
                    config.callLifecycle.apply(config, __spread([framework_shared_1.AppLifecycle.error], params));
                };
                config.callLifecycle(framework_shared_1.AppLifecycle.launch);
                this.handelVisibilityChange();
            };
            AppConfig.prototype.componentWillUnmount = function () {
                document.removeEventListener('visibilitychange', this.visibilityChangeEvent);
            };
            AppConfig.prototype.render = function () {
                var ref;
                if (framework_shared_1.isClassComponent(AppComponent) || AppComponent.$$typeof === react_is_1.ForwardRef) {
                    ref = config._instance;
                }
                return React.createElement(AppComponent, __assign({}, this.props, { ref: ref }));
            };
            return AppConfig;
        }(React.Component));
    };
    return createConfig(App);
}
exports.default = createAppConfig;
