// @remax/web/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bootstrapMpa = exports.bootstrap = exports.usePageInstance = exports.useNativeEffect = exports.useQuery = exports.createPageConfig = exports.createAppConfig = exports.usePageEvent = exports.useAppEvent = void 0;
var framework_shared_1 = require("@remax/framework-shared");
Object.defineProperty(exports, "useAppEvent", { enumerable: true, get: function () { return framework_shared_1.useAppEvent; } });
Object.defineProperty(exports, "usePageEvent", { enumerable: true, get: function () { return framework_shared_1.usePageEvent; } });
var createAppConfig_1 = require("./createAppConfig");
Object.defineProperty(exports, "createAppConfig", { enumerable: true, get: function () { return __importDefault(createAppConfig_1).default; } });
var createPageConfig_1 = require("./createPageConfig");
Object.defineProperty(exports, "createPageConfig", { enumerable: true, get: function () { return __importDefault(createPageConfig_1).default; } });
var useQuery_1 = require("./hooks/useQuery");
Object.defineProperty(exports, "useQuery", { enumerable: true, get: function () { return __importDefault(useQuery_1).default; } });
var useNativeEffect_1 = require("./hooks/useNativeEffect");
Object.defineProperty(exports, "useNativeEffect", { enumerable: true, get: function () { return __importDefault(useNativeEffect_1).default; } });
var usePageInstance_1 = require("./hooks/usePageInstance");
Object.defineProperty(exports, "usePageInstance", { enumerable: true, get: function () { return __importDefault(usePageInstance_1).default; } });
var bootstrap_1 = require("./bootstrap");
Object.defineProperty(exports, "bootstrap", { enumerable: true, get: function () { return __importDefault(bootstrap_1).default; } });
var bootstrapMpa_1 = require("./bootstrapMpa");
Object.defineProperty(exports, "bootstrapMpa", { enumerable: true, get: function () { return __importDefault(bootstrapMpa_1).default; } });
__exportStar(require("./api"), exports);
__exportStar(require("./types"), exports);
