// @remax/one/cjs/hostComponents/useWebTouch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var constants_1 = require("../constants");
function useWebTouch(_a) {
    var hoverDuration = _a.hoverDuration, hoverDelay = _a.hoverDelay, onLongTap = _a.onLongTap, onTouchCancel = _a.onTouchCancel, onTouchEnd = _a.onTouchEnd, onTouchMove = _a.onTouchMove, onTouchStart = _a.onTouchStart;
    var _b = __read(React.useState(false), 2), touching = _b[0], setTouching = _b[1];
    var hoveringRef = React.useRef(false);
    var timers = React.useRef([]);
    function executeTimeout(callback, time) {
        var timer = setTimeout(function () {
            callback();
            timers.current.filter(function (t) { return t !== timer; });
        }, time);
        timers.current.push(timer);
    }
    React.useEffect(function () { return function () {
        timers.current.forEach(function (t) { return clearTimeout(t); });
    }; }, []);
    function handleTouchStart(e) {
        hoveringRef.current = true;
        executeTimeout(function () {
            if (hoveringRef.current) {
                setTouching(true);
            }
        }, hoverDelay);
        executeTimeout(function () {
            if (hoveringRef.current && typeof onLongTap === 'function') {
                onLongTap(e);
            }
        }, constants_1.LONG_TAP_DURATION);
        if (typeof onTouchStart === 'function') {
            return onTouchStart(e);
        }
    }
    function handleTouchMove(e) {
        hoveringRef.current = false;
        executeTimeout(function () {
            if (touching) {
                setTouching(false);
            }
        }, hoverDuration);
        if (typeof onTouchMove === 'function') {
            return onTouchMove(e);
        }
    }
    function handleTouchEnd(e) {
        hoveringRef.current = false;
        executeTimeout(function () {
            if (touching) {
                setTouching(false);
            }
        }, hoverDuration);
        if (typeof onTouchEnd === 'function') {
            return onTouchEnd(e);
        }
    }
    function handleTouchCancel(e) {
        hoveringRef.current = false;
        executeTimeout(function () {
            if (touching) {
                setTouching(false);
            }
        }, hoverDuration);
        if (typeof onTouchCancel === 'function') {
            return onTouchCancel(e);
        }
    }
    return [touching, handleTouchStart, handleTouchMove, handleTouchEnd, handleTouchCancel];
}
exports.default = useWebTouch;
