// @remax/ali/cjs/hostComponents/ScrollView/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.props = exports.alias = void 0;
var shared_1 = require("@remax/shared");
exports.alias = {
    id: 'id',
    className: 'class',
    style: 'style',
    scrollX: 'scroll-x',
    scrollY: 'scroll-y',
    upperThreshold: 'upper-threshold',
    lowerThreshold: 'lower-threshold',
    scrollTop: 'scroll-top',
    scrollLeft: 'scroll-left',
    scrollIntoView: 'scroll-into-view',
    scrollWithAnimation: 'scroll-with-animation',
    scrollAnimationDuration: 'scroll-animation-duration',
    enableBackToTop: 'enable-back-to-top',
    trapScroll: 'trap-scroll',
    onScrollToUpper: 'onScrollToUpper',
    disableLowerScroll: 'disable-lower-scroll',
    disableUpperScroll: 'disable-upper-scroll',
    onScrollToLower: 'onScrollToLower',
    onScroll: 'onScroll',
    onTouchStart: 'onTouchStart',
    onTouchMove: 'onTouchMove',
    onTouchEnd: 'onTouchEnd',
    onTouchCancel: 'onTouchCancel',
};
exports.props = shared_1.unique(Object.values(exports.alias));
