// @remax/toutiao/esm/api/index.js
import { promisify } from '@remax/framework-shared';
export var getAppStub = getApp;
export var canIUse = tt.canIUse;
export var getSystemInfoSync = tt.getSystemInfoSync;
export var getSystemInfo = promisify(tt.getSystemInfo);
export var getUpdateManager = tt.getUpdateManager;
export var getLaunchOptionsSync = tt.getLaunchOptionsSync;
export var exitMiniProgram = tt.exitMiniProgram;
export var switchTab = promisify(tt.switchTab);
export var reLaunch = promisify(tt.reLaunch);
export var redirectTo = promisify(tt.redirectTo);
export var navigateTo = promisify(tt.navigateTo);
export var navigateBack = promisify(tt.navigateBack);
export var showToast = promisify(tt.showToast);
export var showModal = promisify(tt.showModal);
export var showLoading = promisify(tt.showLoading);
export var showActionSheet = promisify(tt.showActionSheet);
export var hideToast = promisify(tt.hideToast);
export var hideLoading = promisify(tt.hideLoading);
export var setNavigationBarTitle = tt.setNavigationBarTitle;
export var stopPullDownRefresh = tt.stopPullDownRefresh;
export var startPullDownRefresh = tt.startPullDownRefresh;
export var pageScrollTo = promisify(tt.pageScrollTo);
export var createAnimation = tt.createAnimation;
export var request = promisify(tt.request);
export var downloadFile = promisify(tt.downloadFile);
export var uploadFile = promisify(tt.uploadFile);
export var connectSocket = tt.connectSocket;
export var setStorageSync = tt.setStorageSync;
export var setStorage = promisify(tt.setStorage);
export var removeStorageSync = tt.removeStorageSync;
export var removeStorage = promisify(tt.removeStorage);
export var getStorageSync = tt.getStorageSync;
export var getStorageInfoSync = tt.getStorageInfoSync;
export var getStorageInfo = promisify(tt.getStorageInfo);
export var getStorage = promisify(tt.getStorage);
export var clearStorageSync = tt.clearStorageSync;
export var clearStorage = promisify(tt.clearStorage);
export var saveImageToPhotosAlbum = promisify(tt.saveImageToPhotosAlbum);
export var previewImage = promisify(tt.previewImage);
export var getImageInfo = promisify(tt.getImageInfo);
export var chooseImage = promisify(tt.chooseImage);
export var saveVideoToPhotosAlbum = promisify(tt.saveVideoToPhotosAlbum);
export var createVideoContext = tt.createVideoContext;
export var chooseVideo = promisify(tt.chooseVideo);
export var openLocation = promisify(tt.openLocation);
export var getLocation = promisify(tt.getLocation);
export var showShareMenu = promisify(tt.showShareMenu);
export var hideShareMenu = promisify(tt.hideShareMenu);
export var navigateToVideoView = promisify(tt.navigateToVideoView);
export var createCanvasContext = tt.createCanvasContext;
export var saveFile = promisify(tt.saveFile);
export var removeSavedFile = promisify(tt.removeSavedFile);
export var openDocument = promisify(tt.openDocument);
export var getSavedFileList = tt.getSavedFileList;
export var getFileSystemManager = tt.getFileSystemManager;
export var getFileInfo = tt.getFileInfo;
export var login = promisify(tt.login);
export var checkSession = tt.checkSession;
export var navigateToMiniProgram = tt.navigateToMiniProgram;
export var navigateBackMiniProgram = promisify(tt.navigateBackMiniProgram);
export var getUserInfo = promisify(tt.getUserInfo);
export var reportAnalytics = tt.reportAnalytics;
export var requestPayment = promisify(tt.requestPayment);
export var pay = promisify(tt.pay);
export var authorize = promisify(tt.authorize);
export var openSetting = promisify(tt.openSetting);
export var getSetting = promisify(tt.getSetting);
export var chooseAddress = promisify(tt.chooseAddress);
export var setClipboardData = promisify(tt.setClipboardData);
export var getClipboardData = tt.getClipboardData;
export var onNetworkStatusChange = tt.onNetworkStatusChange;
export var getNetworkType = promisify(tt.getNetworkType);
export var makePhoneCall = promisify(tt.makePhoneCall);
export var stopAccelerometer = promisify(tt.stopAccelerometer);
export var startAccelerometer = promisify(tt.startAccelerometer);
export var onAccelerometerChange = tt.onAccelerometerChange;
export var stopCompass = promisify(tt.stopCompass);
export var startCompass = promisify(tt.startCompass);
export var onCompassChange = tt.onCompassChange;
export var scanCode = promisify(tt.scanCode);
export var vibrateShort = promisify(tt.vibrateShort);
export var vibrateLong = promisify(tt.vibrateLong);
export var getExtConfigSync = tt.getExtConfigSync;
export var getExtConfig = promisify(tt.getExtConfig);
export var createSelectorQuery = tt.createSelectorQuery;
export var createIntersectionObserver = tt.createIntersectionObserver;
export var createRewardedVideoAd = tt.createRewardedVideoAd;
export var getConnectedWifi = promisify(tt.getConnectedWifi);
export var setKeepScreenOn = promisify(tt.setKeepScreenOn);
export var getMenuButtonLayout = tt.getMenuButtonLayout;
export var getMenuButtonBoundingClientRect = tt.getMenuButtonBoundingClientRect;
export var createInnerAudioContext = tt.createInnerAudioContext;
export var getBackgroundAudioManager = tt.getBackgroundAudioManager;
export var getRecorderManager = tt.getRecorderManager;
export var createLivePlayerContext = tt.createLivePlayerContext;
export var createMapContext = tt.createMapContext;
export var interactionAnalysis = promisify(tt.interactionAnalysis);
export var followOfficialAccount = promisify(tt.followOfficialAccount);
export var checkFollowState = promisify(tt.checkFollowState);
export var onMemoryWarning = tt.onMemoryWarning;
export var showTabBarRedDot = promisify(tt.showTabBarRedDot);
export var showTabBar = promisify(tt.showTabBar);
export var setTabBarStyle = promisify(tt.setTabBarStyle);
export var setTabBarItem = promisify(tt.setTabBarItem);
export var setTabBarBadge = promisify(tt.setTabBarBadge);
export var removeTabBarBadge = promisify(tt.removeTabBarBadge);
export var hideTabBarRedDot = promisify(tt.hideTabBarRedDot);
export var hideTabBar = promisify(tt.hideTabBar);
