// @remax/ali/cjs/hostComponents/Swiper/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.props = exports.alias = void 0;
var shared_1 = require("@remax/shared");
exports.alias = {
    id: 'id',
    className: 'class',
    style: 'style',
    indicatorDots: 'indicator-dots',
    indicatorColor: 'indicator-color',
    indicatorActiveColor: 'indicator-active-color',
    activeClass: 'active-class',
    activeClassName: 'active-class',
    changingClass: 'changing-class',
    changingClassName: 'changing-class',
    autoplay: 'autoplay',
    current: 'current',
    duration: 'duration',
    interval: 'interval',
    circular: 'circular',
    vertical: 'vertical',
    previousMargin: 'previous-margin',
    nextMargin: 'next-margin',
    acceleration: 'acceleration',
    disableProgrammaticAnimation: 'disable-programmatic-animation',
    onChange: 'onChange',
    onTransition: 'onTransition',
    onAnimationEnd: 'onAnimationEnd',
    disableTouch: 'disable-touch',
};
exports.props = shared_1.unique(Object.values(exports.alias));
