// @remax/ali/cjs/hostComponents/View/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.props = exports.alias = void 0;
var shared_1 = require("@remax/shared");
exports.alias = {
    id: 'id',
    disableScroll: 'disable-scroll',
    hoverClass: 'hover-class',
    hoverClassName: 'hover-class',
    hoverStartTime: 'hover-start-time',
    hoverStayTime: 'hover-stay-time',
    hidden: 'hidden',
    className: 'class',
    style: 'style',
    animation: 'animation',
    hoverStopPropagation: 'hover-stop-propagation',
    onClick: 'onTap',
    onTap: 'onTap',
    onTouchStart: 'onTouchStart',
    onTouchMove: 'onTouchMove',
    onTouchEnd: 'onTouchEnd',
    onTouchCancel: 'onTouchCancel',
    onLongTap: 'onLongTap',
    onLongClick: 'onLongTap',
    onTransitionEnd: 'onTransitionEnd',
    onAnimationIteration: 'onAnimationIteration',
    onAnimationStart: 'onAnimationStart',
    onAnimationEnd: 'onAnimationEnd',
    onAppear: 'onAppear',
    onDisappear: 'onDisappear',
    onFirstAppear: 'onFirstAppear',
};
exports.props = shared_1.unique(Object.values(exports.alias));
