// @remax/runtime/cjs/AppContainer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var VNode_1 = __importDefault(require("./VNode"));
var instanceId_1 = require("./instanceId");
var AppContainer = /** @class */ (function () {
    function AppContainer() {
        this.updateQueue = [];
        this.root = new VNode_1.default({
            id: instanceId_1.generate(),
            type: 'root',
            container: this,
        });
        this.root.mounted = true;
    }
    AppContainer.prototype.requestUpdate = function (path, start, deleteCount) {
        var items = [];
        for (var _i = 3; _i < arguments.length; _i++) {
            items[_i - 3] = arguments[_i];
        }
        // ignore
    };
    AppContainer.prototype.applyUpdate = function () {
        this.context._pages.forEach(function (page) {
            page.container.applyUpdate();
            page.modalContainer.applyUpdate();
        });
    };
    AppContainer.prototype.createCallback = function (name, fn) {
        this.context[name] = fn;
    };
    AppContainer.prototype.appendChild = function (child) {
        this.root.appendChild(child);
    };
    AppContainer.prototype.removeChild = function (child) {
        this.root.removeChild(child);
    };
    AppContainer.prototype.insertBefore = function (child, beforeChild) {
        this.root.insertBefore(child, beforeChild);
    };
    return AppContainer;
}());
exports.default = AppContainer;
