// @remax/runtime/cjs/createPageConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetPageId = void 0;
var React = __importStar(require("react"));
var framework_shared_1 = require("@remax/framework-shared");
var stopPullDownRefresh_1 = __importDefault(require("./stopPullDownRefresh"));
var Container_1 = __importDefault(require("./Container"));
var ReactPortal_1 = require("./ReactPortal");
var render_1 = __importDefault(require("./render"));
var idCounter = 0;
function generatePageId() {
    var id = idCounter;
    var pageId = 'page_' + id;
    idCounter += 1;
    return pageId;
}
// for testing
function resetPageId() {
    idCounter = 0;
}
exports.resetPageId = resetPageId;
function createPageConfig(Page, name) {
    var _a;
    var app = getApp();
    var config = {
        data: {
            root: {
                children: [],
            },
            modalRoot: {
                children: [],
            },
        },
        wrapperRef: React.createRef(),
        lifecycleCallback: {},
        onLoad: function (query) {
            var PageWrapper = framework_shared_1.createPageWrapper(Page, name);
            this.pageId = generatePageId();
            this.lifecycleCallback = {};
            this.data = {
                root: {
                    children: [],
                },
                modalRoot: {
                    children: [],
                },
            };
            this.query = query;
            this.container = new Container_1.default(this, 'root');
            this.modalContainer = new Container_1.default(this, 'modalRoot');
            var pageElement = React.createElement(PageWrapper, {
                page: this,
                query: query,
                modalContainer: this.modalContainer,
                ref: this.wrapperRef,
            });
            if (app && app._mount) {
                this.element = ReactPortal_1.createPortal(pageElement, this.container, this.pageId);
                app._mount(this);
            }
            else {
                this.element = render_1.default(pageElement, this.container);
            }
            return this.callLifecycle(framework_shared_1.Lifecycle.load, query);
        },
        onUnload: function () {
            this.callLifecycle(framework_shared_1.Lifecycle.unload);
            this.unloaded = true;
            this.container.clearUpdate();
            app._unmount(this);
        },
        onTabItemTap: function (e) {
            return this.callLifecycle(framework_shared_1.Lifecycle.tabItemTap, e);
        },
        onResize: function (e) {
            return this.callLifecycle(framework_shared_1.Lifecycle.resize, e);
        },
        /**
         * Lifecycle start
         */
        registerLifecycle: function (lifecycle, callback) {
            var _this = this;
            this.lifecycleCallback[lifecycle] = this.lifecycleCallback[lifecycle] || [];
            this.lifecycleCallback[lifecycle].push(callback);
            return function () {
                _this.lifecycleCallback[lifecycle].splice(_this.lifecycleCallback[lifecycle].indexOf(callback), 1);
            };
        },
        callLifecycle: function (lifecycle) {
            var _a;
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            var callbacks = this.lifecycleCallback[lifecycle] || [];
            var result;
            callbacks.forEach(function (callback) {
                result = callback.apply(void 0, __spread(args));
            });
            if (result) {
                return result;
            }
            var callback = framework_shared_1.callbackName(lifecycle);
            if (this.wrapperRef.current && this.wrapperRef.current[callback]) {
                return (_a = this.wrapperRef.current)[callback].apply(_a, __spread(args));
            }
        },
        events: {
            // 页面返回时触发
            onBack: function (e) {
                return this.callLifecycle(framework_shared_1.Lifecycle.back, e);
            },
            // 键盘高度变化时触发
            onKeyboardHeight: function (e) {
                return this.callLifecycle(framework_shared_1.Lifecycle.keyboardHeight, e);
            },
            onTabItemTap: function (e) {
                return this.callLifecycle(framework_shared_1.Lifecycle.tabItemTap, e);
            },
            // 点击但切换tabItem前触发
            beforeTabItemTap: function () {
                return this.callLifecycle(framework_shared_1.Lifecycle.beforeTabItemTap);
            },
            onResize: function (e) {
                return this.callLifecycle(framework_shared_1.Lifecycle.resize, e);
            },
        },
        onShow: function () {
            return this.callLifecycle(framework_shared_1.Lifecycle.show);
        },
        onHide: function () {
            return this.callLifecycle(framework_shared_1.Lifecycle.hide);
        },
        onReady: function () {
            return this.callLifecycle(framework_shared_1.Lifecycle.ready);
        },
        onPullDownRefresh: function (e) {
            var result = this.callLifecycle(framework_shared_1.Lifecycle.pullDownRefresh, e);
            if (result && typeof result.then === 'function') {
                Promise.resolve(result).then(function () {
                    stopPullDownRefresh_1.default();
                });
            }
        },
        onReachBottom: function () {
            return this.callLifecycle(framework_shared_1.Lifecycle.reachBottom);
        },
        onTitleClick: function () {
            return this.callLifecycle(framework_shared_1.Lifecycle.titleClick);
        },
        onOptionMenuClick: function (e) {
            return this.callLifecycle(framework_shared_1.Lifecycle.optionMenuClick, e);
        },
        onPopMenuClick: function (e) {
            return this.callLifecycle(framework_shared_1.Lifecycle.popMenuClick, e);
        },
        onPullIntercept: function () {
            return this.callLifecycle(framework_shared_1.Lifecycle.pullIntercept);
        },
    };
    var lifecycleEvents = {
        onPageScroll: function (e) {
            return this.callLifecycle(framework_shared_1.Lifecycle.pageScroll, e);
        },
        onShareAppMessage: function (options) {
            return this.callLifecycle(framework_shared_1.Lifecycle.shareAppMessage, options) || {};
        },
        onShareTimeline: function (options) {
            return this.callLifecycle(framework_shared_1.Lifecycle.shareTimeline, options) || {};
        },
    };
    ((_a = framework_shared_1.RuntimeOptions.get('pageEvents')[name]) !== null && _a !== void 0 ? _a : []).forEach(function (eventName) {
        if (lifecycleEvents[eventName]) {
            config[eventName] = lifecycleEvents[eventName];
        }
    });
    return framework_shared_1.RuntimeOptions.get('pluginDriver').onPageConfig({ config: config, page: name });
}
exports.default = createPageConfig;
