// @remax/runtime/cjs/createComponentConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var framework_shared_1 = require("@remax/framework-shared");
var Container_1 = __importDefault(require("./Container"));
var render_1 = __importDefault(require("./render"));
function createComponentConfig(Component) {
    var config = {
        data: {
            action: {},
            root: {
                children: [],
            },
        },
        didMount: function () {
            if (!this.container) {
                this.init();
            }
        },
        didUpdate: function (prevProps, prevData) {
            if (prevData !== this.data) {
                return;
            }
            this.render();
        },
        didUnmount: function () {
            this.container.clearUpdate();
            render_1.default(null, this.container);
        },
        methods: {
            init: function () {
                this.component = framework_shared_1.RuntimeOptions.get('pluginDriver').onMiniComponent({
                    component: Component,
                    context: this,
                });
                this.container = new Container_1.default(this);
                this.render();
            },
            render: function () {
                this.element = render_1.default(React.createElement(framework_shared_1.ComponentInstanceContext.Provider, {
                    value: this,
                }, React.createElement(this.component, this.props)), this.container);
            },
        },
    };
    return config;
}
exports.default = createComponentConfig;
