// @remax/runtime/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unstable_batchedUpdates = exports.PluginDriver = exports.RuntimeOptions = exports.createPortal = exports.createHostComponent = exports.createNativeComponent = exports.createComponentConfig = exports.createPageConfig = exports.createAppConfig = exports.render = void 0;
var render_1 = require("./render");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return __importDefault(render_1).default; } });
var createAppConfig_1 = require("./createAppConfig");
Object.defineProperty(exports, "createAppConfig", { enumerable: true, get: function () { return __importDefault(createAppConfig_1).default; } });
var createPageConfig_1 = require("./createPageConfig");
Object.defineProperty(exports, "createPageConfig", { enumerable: true, get: function () { return __importDefault(createPageConfig_1).default; } });
var createComponentConfig_1 = require("./createComponentConfig");
Object.defineProperty(exports, "createComponentConfig", { enumerable: true, get: function () { return __importDefault(createComponentConfig_1).default; } });
var createNativeComponent_1 = require("./createNativeComponent");
Object.defineProperty(exports, "createNativeComponent", { enumerable: true, get: function () { return __importDefault(createNativeComponent_1).default; } });
var createHostComponent_1 = require("./createHostComponent");
Object.defineProperty(exports, "createHostComponent", { enumerable: true, get: function () { return __importDefault(createHostComponent_1).default; } });
var ReactPortal_1 = require("./ReactPortal");
Object.defineProperty(exports, "createPortal", { enumerable: true, get: function () { return ReactPortal_1.createPortal; } });
var framework_shared_1 = require("@remax/framework-shared");
Object.defineProperty(exports, "RuntimeOptions", { enumerable: true, get: function () { return framework_shared_1.RuntimeOptions; } });
Object.defineProperty(exports, "PluginDriver", { enumerable: true, get: function () { return framework_shared_1.PluginDriver; } });
__exportStar(require("./hooks"), exports);
var render_2 = require("./render");
exports.unstable_batchedUpdates = render_2.ReactReconcilerInst.batchedUpdates;
exports.default = {
    unstable_batchedUpdates: exports.unstable_batchedUpdates,
};
