// @remax/cli/lib/build/webpack/config.web.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const webpack_chain_1 = __importDefault(require("webpack-chain"));
const html_webpack_plugin_1 = __importDefault(require("html-webpack-plugin"));
const baseConfig_1 = __importDefault(require("./baseConfig"));
const webBaseConfig_1 = __importDefault(require("./webBaseConfig"));
const SpaEntry_1 = __importDefault(require("../entries/SpaEntry"));
function webpackConfig(builder) {
    const config = new webpack_chain_1.default();
    config.output.publicPath('/');
    baseConfig_1.default(config, builder);
    const spaEntry = new SpaEntry_1.default(builder, 'index', './_index.js');
    config.entry(spaEntry.name).add(spaEntry.virtualPath);
    config.plugin('webpack-virtual-modules').use(spaEntry.virtualModule);
    config.optimization.splitChunks({
        cacheGroups: {
            remaxStyles: {
                name: 'remax-styles',
                test: new RegExp(`(.css|.less|.sass|.scss|.stylus|.styl|${builder.api.meta.style})$`),
                chunks: 'all',
                minChunks: 2,
                minSize: 0,
            },
        },
    });
    config.plugin('html-webpack-plugin').use(html_webpack_plugin_1.default, [
        {
            template: fs.existsSync(path.join(builder.projectPath.publicDir(), '/index.html'))
                ? path.join(builder.projectPath.publicDir(), '/index.html')
                : path.resolve(__dirname, '../../../template/index.html.ejs'),
            env: process.env.NODE_ENV,
        },
    ]);
    webBaseConfig_1.default(config, builder);
    return config.toConfig();
}
exports.default = webpackConfig;
