// @remax/cli/lib/build/webpack/plugins/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeAsset = exports.PluginAsset = exports.CoverageIgnore = exports.RuntimeOptions = exports.OptimizeEntries = exports.ComponentAsset = exports.PageAsset = exports.AppAsset = void 0;
var AppAsset_1 = require("./AppAsset");
Object.defineProperty(exports, "AppAsset", { enumerable: true, get: function () { return __importDefault(AppAsset_1).default; } });
var PageAsset_1 = require("./PageAsset");
Object.defineProperty(exports, "PageAsset", { enumerable: true, get: function () { return __importDefault(PageAsset_1).default; } });
var ComponentAsset_1 = require("./ComponentAsset");
Object.defineProperty(exports, "ComponentAsset", { enumerable: true, get: function () { return __importDefault(ComponentAsset_1).default; } });
var OptimizeEntries_1 = require("./OptimizeEntries");
Object.defineProperty(exports, "OptimizeEntries", { enumerable: true, get: function () { return __importDefault(OptimizeEntries_1).default; } });
var RuntimeOptions_1 = require("./RuntimeOptions");
Object.defineProperty(exports, "RuntimeOptions", { enumerable: true, get: function () { return __importDefault(RuntimeOptions_1).default; } });
var CoverageIgnore_1 = require("./CoverageIgnore");
Object.defineProperty(exports, "CoverageIgnore", { enumerable: true, get: function () { return __importDefault(CoverageIgnore_1).default; } });
var PluginAsset_1 = require("./PluginAsset");
Object.defineProperty(exports, "PluginAsset", { enumerable: true, get: function () { return __importDefault(PluginAsset_1).default; } });
var NativeAsset_1 = require("./NativeAsset");
Object.defineProperty(exports, "NativeAsset", { enumerable: true, get: function () { return __importDefault(NativeAsset_1).default; } });
