// @remax/one/cjs/createHostComponent.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.aliasProps = exports.createFormEvent = exports.createInputEvent = exports.createCallback = exports.createImageEvent = exports.createTouchEvent = exports.createTapEvent = exports.createCurrentTarget = exports.createTarget = void 0;
var React = __importStar(require("react"));
var memoize_one_1 = __importDefault(require("memoize-one"));
var framework_shared_1 = require("@remax/framework-shared");
function createTarget(target, detail) {
    return {
        id: target.id,
        offsetLeft: target.offsetLeft,
        offsetTop: target.offsetTop,
        dataset: target.targetDataset || target.dataset,
        value: detail === null || detail === void 0 ? void 0 : detail.value,
    };
}
exports.createTarget = createTarget;
function createCurrentTarget(currentTarget) {
    return {
        id: currentTarget.id,
        offsetLeft: currentTarget.offsetLeft,
        offsetTop: currentTarget.offsetTop,
        dataset: currentTarget.dataset,
    };
}
exports.createCurrentTarget = createCurrentTarget;
var createTapEvent = function (originalEvent) { return ({
    type: originalEvent.type,
    stopPropagation: originalEvent.stopPropagation,
    target: createTarget(originalEvent.target, originalEvent.detail),
    currentTarget: createCurrentTarget(originalEvent.currentTarget),
    originalEvent: originalEvent,
    nativeEvent: originalEvent,
}); };
exports.createTapEvent = createTapEvent;
var createTouchEvent = function (originalEvent) { return ({
    type: originalEvent.type,
    stopPropagation: originalEvent.stopPropagation,
    target: createTarget(originalEvent.target, originalEvent.detail),
    currentTarget: createCurrentTarget(originalEvent.currentTarget),
    touches: originalEvent.touches,
    changedTouches: originalEvent.touches,
    originalEvent: originalEvent,
    nativeEvent: originalEvent,
}); };
exports.createTouchEvent = createTouchEvent;
var createImageEvent = function (originalEvent) { return ({
    type: originalEvent.type,
    target: createTarget(originalEvent.target, originalEvent.detail),
    currentTarget: createCurrentTarget(originalEvent.currentTarget),
    originalEvent: originalEvent,
    nativeEvent: originalEvent,
}); };
exports.createImageEvent = createImageEvent;
function createCallback(fn, eventCreator) {
    if (typeof fn !== 'function') {
        return undefined;
    }
    return function (originalEvent) { return fn(eventCreator(originalEvent)); };
}
exports.createCallback = createCallback;
var createInputEvent = function (originalEvent) { return ({
    type: originalEvent.type,
    target: createTarget(originalEvent.target, originalEvent.detail),
    currentTarget: createCurrentTarget(originalEvent.currentTarget),
    originalEvent: originalEvent,
    nativeEvent: originalEvent,
}); };
exports.createInputEvent = createInputEvent;
var createFormEvent = function (originalEvent) { return ({
    type: originalEvent.type,
    target: createTarget(originalEvent.target, originalEvent.detail),
    currentTarget: createCurrentTarget(originalEvent.currentTarget),
    originalEvent: originalEvent,
    nativeEvent: originalEvent,
}); };
exports.createFormEvent = createFormEvent;
function assignDefaultProps(inputProps, defaultProps) {
    if (defaultProps) {
        Object.keys(defaultProps).forEach(function (key) {
            var _a;
            inputProps[key] = (_a = inputProps[key]) !== null && _a !== void 0 ? _a : defaultProps[key];
        });
    }
}
function aliasProps(props, alias) {
    var _a;
    if (!alias) {
        return props;
    }
    var nextProps = {};
    for (var key in props) {
        nextProps[(_a = alias[key]) !== null && _a !== void 0 ? _a : key] = props[key];
    }
    return nextProps;
}
exports.aliasProps = aliasProps;
var createLongTapCallback = memoize_one_1.default(createCallback);
var createTapCallback = memoize_one_1.default(createCallback);
var createTouchStartCallback = memoize_one_1.default(createCallback);
var createTouchMoveCallback = memoize_one_1.default(createCallback);
var createTouchEndCallback = memoize_one_1.default(createCallback);
var createTouchCancelCallback = memoize_one_1.default(createCallback);
var createChangeCallback = memoize_one_1.default(createCallback);
var createInputCallback = memoize_one_1.default(createCallback);
var createConfirmCallback = memoize_one_1.default(createCallback);
var createFocusCallback = memoize_one_1.default(createCallback);
var createBlurCallback = memoize_one_1.default(createCallback);
var createSubmitCallback = memoize_one_1.default(createCallback);
var createResetCallback = memoize_one_1.default(createCallback);
var createImageLoadCallback = memoize_one_1.default(createCallback);
var createImageErrorCallback = memoize_one_1.default(createCallback);
function createHostComponent(name, alias, defaults) {
    var Component = function (props, ref) {
        var inputProps = __assign({}, props);
        // 默认属性根据平台在这里设置
        if (defaults) {
            assignDefaultProps(inputProps, defaults);
        }
        if (props.onLongTap) {
            inputProps.onLongTap = createLongTapCallback(inputProps.onLongTap, exports.createTapEvent);
        }
        if (inputProps.onTap) {
            inputProps.onTap = createTapCallback(inputProps.onTap, exports.createTapEvent);
        }
        if (inputProps.onTouchStart) {
            inputProps.onTouchStart = createTouchStartCallback(inputProps.onTouchStart, exports.createTouchEvent);
        }
        if (inputProps.onTouchMove) {
            inputProps.onTouchMove = createTouchMoveCallback(inputProps.onTouchMove, exports.createTouchEvent);
        }
        if (inputProps.onTouchEnd) {
            inputProps.onTouchEnd = createTouchEndCallback(inputProps.onTouchEnd, exports.createTouchEvent);
        }
        if (inputProps.onTouchCancel) {
            inputProps.onTouchCancel = createTouchCancelCallback(inputProps.onTouchCancel, exports.createTouchEvent);
        }
        if (inputProps.onChange) {
            inputProps.onChange = createChangeCallback(inputProps.onChange, exports.createInputEvent);
        }
        if (inputProps.onInput) {
            inputProps.onInput = createInputCallback(inputProps.onInput, exports.createInputEvent);
        }
        if (inputProps.onConfirm) {
            inputProps.onConfirm = createConfirmCallback(inputProps.onConfirm, exports.createInputEvent);
        }
        if (inputProps.onFocus) {
            inputProps.onFocus = createFocusCallback(inputProps.onFocus, exports.createInputEvent);
        }
        if (inputProps.onBlur) {
            inputProps.onBlur = createBlurCallback(inputProps.onBlur, exports.createInputEvent);
        }
        if (inputProps.onSubmit) {
            inputProps.onSubmit = createSubmitCallback(inputProps.onSubmit, exports.createFormEvent);
        }
        if (inputProps.onReset) {
            inputProps.onReset = createResetCallback(inputProps.onReset, exports.createFormEvent);
        }
        if (name === 'image') {
            if (inputProps.onLoad) {
                inputProps.onLoad = createImageLoadCallback(props.onLoad, exports.createImageEvent);
            }
            if (inputProps.onError) {
                inputProps.onError = createImageErrorCallback(props.onError, exports.createImageEvent);
            }
        }
        var nextProps = inputProps;
        if (alias) {
            nextProps = aliasProps(inputProps, alias);
        }
        nextProps.ref = ref;
        return React.createElement(name, nextProps);
    };
    if (process.env.NODE_ENV === 'development') {
        Component.displayName = framework_shared_1.formatDisplayName(name);
    }
    return React.forwardRef(Component);
}
exports.default = createHostComponent;
