// @remax/plugin-devtools/lib/WebSocket.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const global_1 = __importDefault(require("./global"));
var ReadyState;
(function (ReadyState) {
    ReadyState[ReadyState["CONNECTING"] = 0] = "CONNECTING";
    ReadyState[ReadyState["OPEN"] = 1] = "OPEN";
    ReadyState[ReadyState["CLOSING"] = 2] = "CLOSING";
    ReadyState[ReadyState["CLOSED"] = 3] = "CLOSED";
})(ReadyState || (ReadyState = {}));
class WebSocket {
    constructor(url) {
        this.CONNECTING = ReadyState.CONNECTING;
        this.OPEN = ReadyState.OPEN;
        this.CLOSING = ReadyState.CLOSING;
        this.CLOSED = ReadyState.CLOSED;
        this.readyState = ReadyState.CONNECTING;
        this.ws = global_1.default.connectSocket({
            url,
        });
        this.ws.onOpen(() => {
            this.readyState = ReadyState.OPEN;
            if (typeof this.onopen === 'function') {
                this.onopen();
            }
        });
        this.ws.onError((res) => {
            if (typeof this.onerror === 'function') {
                this.onerror(res);
            }
        });
        this.ws.onClose(() => {
            this.readyState = ReadyState.CLOSED;
            if (typeof this.onclose === 'function') {
                this.onclose();
            }
        });
        this.ws.onMessage((res) => {
            if (typeof this.onmessage === 'function') {
                this.onmessage(res);
            }
        });
    }
    send(payload) {
        this.ws.send({
            data: payload,
        });
    }
}
exports.default = WebSocket;
