// @remax/framework-shared/cjs/createPageWrapper.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var react_is_1 = require("react-is");
var isClassComponent_1 = __importDefault(require("./utils/isClassComponent"));
var lifecycle_1 = require("./lifecycle");
var PageInstanceContext_1 = __importDefault(require("./PageInstanceContext"));
var RuntimeOptions = __importStar(require("./RuntimeOptions"));
function createPageWrapper(Page, name) {
    var WrappedPage = RuntimeOptions.get('pluginDriver').onPageComponent({ component: Page, page: name });
    return /** @class */ (function (_super) {
        __extends(PageWrapper, _super);
        function PageWrapper(props) {
            var _this = _super.call(this, props) || this;
            // 页面组件的实例
            _this.pageComponentInstance = null;
            _this.callbacks = new Map();
            Object.keys(lifecycle_1.Lifecycle).forEach(function (phase) {
                var callback = lifecycle_1.callbackName(phase);
                _this[callback] = function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i] = arguments[_i];
                    }
                    return _this.callLifecycle.apply(_this, __spread([phase], args));
                };
            });
            return _this;
        }
        PageWrapper.prototype.callLifecycle = function (phase) {
            var _a;
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            var callback = lifecycle_1.callbackName(phase);
            if (this.pageComponentInstance && typeof this.pageComponentInstance[callback] === 'function') {
                return (_a = this.pageComponentInstance)[callback].apply(_a, __spread(args));
            }
        };
        PageWrapper.prototype.render = function () {
            var _this = this;
            var props = {
                location: {
                    query: this.props.query || {},
                },
            };
            if (isClassComponent_1.default(Page) || Page.$$typeof === react_is_1.ForwardRef) {
                props.ref = function (node) { return (_this.pageComponentInstance = node); };
            }
            return React.createElement(PageInstanceContext_1.default.Provider, { value: this.props.page }, React.createElement(WrappedPage, props));
        };
        return PageWrapper;
    }(React.Component));
}
exports.default = createPageWrapper;
