// @remax/framework-shared/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isClassComponent = exports.PluginDriver = exports.createPageWrapper = exports.ComponentInstanceContext = exports.PageInstanceContext = exports.AppInstanceContext = exports.RuntimeOptions = void 0;
var RuntimeOptions = __importStar(require("./RuntimeOptions"));
exports.RuntimeOptions = RuntimeOptions;
var AppInstanceContext_1 = require("./AppInstanceContext");
Object.defineProperty(exports, "AppInstanceContext", { enumerable: true, get: function () { return __importDefault(AppInstanceContext_1).default; } });
var PageInstanceContext_1 = require("./PageInstanceContext");
Object.defineProperty(exports, "PageInstanceContext", { enumerable: true, get: function () { return __importDefault(PageInstanceContext_1).default; } });
var ComponentInstanceContext_1 = require("./ComponentInstanceContext");
Object.defineProperty(exports, "ComponentInstanceContext", { enumerable: true, get: function () { return __importDefault(ComponentInstanceContext_1).default; } });
var createPageWrapper_1 = require("./createPageWrapper");
Object.defineProperty(exports, "createPageWrapper", { enumerable: true, get: function () { return __importDefault(createPageWrapper_1).default; } });
var PluginDriver_1 = require("./PluginDriver");
Object.defineProperty(exports, "PluginDriver", { enumerable: true, get: function () { return __importDefault(PluginDriver_1).default; } });
var isClassComponent_1 = require("./utils/isClassComponent");
Object.defineProperty(exports, "isClassComponent", { enumerable: true, get: function () { return __importDefault(isClassComponent_1).default; } });
__exportStar(require("./lifecycle"), exports);
__exportStar(require("./hooks"), exports);
__exportStar(require("./formatDisplayName"), exports);
__exportStar(require("./promisify"), exports);
