// @remax/cli/lib/build/entries/MpaEntry.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const VirtualEntry_1 = __importDefault(require("./VirtualEntry"));
class MpaEntry extends VirtualEntry_1.default {
    outputSource() {
        const config = this.builder.api.onPageConfig({
            page: this.name,
            config: this.getManifest(),
        });
        return `
      import App from '@/app';
      import Page from './${path.basename(this.filename)}';
      import { bootstrapMpa } from '@remax/web';
      import '@remax/web/assets/normalize.css';
      import '@remax/web/assets/app.css';

      bootstrapMpa({
        appComponent: App,
        pageComponent: Page,
        page: ${JSON.stringify({
            route: this.name,
            config,
        }, null, 2)},
        appConfig: ${JSON.stringify(this.builder.projectConfig, null, 2)},
        plugins: [
          ${this.builder.api
            .getRuntimePluginFiles()
            .map(file => `require('${file}')`)
            .join(',')}
        ]
      });
    `;
    }
}
exports.default = MpaEntry;
