// @remax/cli/lib/build/entries/SpaEntry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const VirtualEntry_1 = __importDefault(require("./VirtualEntry"));
class SpaEntry extends VirtualEntry_1.default {
    outputSource() {
        return `
      import App from '@/app';
      import { bootstrap } from '@remax/web';
      import '@remax/web/assets/normalize.css';
      import '@remax/web/assets/app.css';

      bootstrap({
        appComponent: App,
        appConfig: ${JSON.stringify(this.builder.projectConfig, null, 2)},
        pages: ${JSON.stringify(Array.from(this.builder.entryCollection.entries.values()).map(entry => {
            const config = this.builder.api.onPageConfig({
                page: entry.name,
                config: entry.getManifest(),
            });
            return {
                route: entry.name,
                config,
            };
        }), null, 2)},
        pageComponents: [
          ${Array.from(this.builder.entryCollection.entries.values())
            .map(entry => `function() {
  return import(
    /* webpackPrefetch: true */
    /* webpackChunkName: "${entry.name}" */
    '${entry.filename}'
  )
}`)
            .join(',')}
        ],
        plugins: [
          ${this.builder.api
            .getRuntimePluginFiles()
            .map(file => `require('${file}')`)
            .join(',')}
        ]
      });
    `;
    }
}
exports.default = SpaEntry;
