// @remax/ali/cjs/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSetting = exports.getServerTime = exports.getScreenBrightness = exports.getSavedFileList = exports.getSavedFileInfo = exports.getRunScene = exports.getRunData = exports.getPhoneNumber = exports.getNetworkType = exports.getLocation = exports.getImageInfo = exports.getFileInfo = exports.getConnectedBluetoothDevices = exports.getClipboard = exports.getBluetoothDevices = exports.getBluetoothAdapterState = exports.getBLEDeviceServices = exports.getBLEDeviceCharacteristics = exports.getBeacons = exports.getBatteryInfoSync = exports.getBatteryInfo = exports.getAuthUserInfo = exports.getAuthCode = exports.downloadFile = exports.disconnectBLEDevice = exports.datePicker = exports.createWebViewContext = exports.createSelectorQuery = exports.createMapContext = exports.createIntersectionObserver = exports.createCanvasContext = exports.createAnimation = exports.connectSocket = exports.connectBLEDevice = exports.confirm = exports.compressImage = exports.closeSocket = exports.closeBluetoothAdapter = exports.clearStorageSync = exports.clearStorage = exports.choosePhoneContact = exports.chooseLocation = exports.chooseImage = exports.chooseContact = exports.chooseCity = exports.chooseAlipayContact = exports.canIUse = exports.alert = exports.addPhoneContact = exports.addCardAuth = void 0;
exports.onCompassChange = exports.onBluetoothDeviceFound = exports.onBluetoothAdapterStateChange = exports.onBLEConnectionStateChanged = exports.onBLECharacteristicValueChange = exports.onBeaconUpdate = exports.onBeaconServiceChange = exports.onAccelerometerChange = exports.offUserCaptureScreen = exports.offSocketOpen = exports.offSocketMessage = exports.offSocketError = exports.offSocketClose = exports.offNetworkStatusChange = exports.offMemoryWarning = exports.offGyroscopeChange = exports.offCompassChange = exports.offBluetoothDeviceFound = exports.offBluetoothAdapterStateChange = exports.offBLEConnectionStateChanged = exports.offBLECharacteristicValueChange = exports.offAccelerometerChange = exports.notifyBLECharacteristicValueChange = exports.navigateToMiniProgram = exports.navigateTo = exports.navigateBackMiniProgram = exports.navigateBack = exports.multiLevelSelect = exports.makePhoneCall = exports.loadFontFace = exports.hideToast = exports.hideTabBarRedDot = exports.hideTabBar = exports.hideShareMenu = exports.hideNavigationBarLoading = exports.hideLoading = exports.hideKeyboard = exports.hideFavoriteMenu = exports.hideBackHome = exports.hideAllFavoriteMenu = exports.hideAllAddToDesktopMenu = exports.hideAddToDesktopMenu = exports.getUpdateManager = exports.getTitleColor = exports.getSystemInfoSync = exports.getSystemInfo = exports.getStorageSync = exports.getStorageInfoSync = exports.getStorageInfo = exports.getStorage = void 0;
exports.setStorageSync = exports.setStorage = exports.setScreenBrightness = exports.setOptionMenu = exports.setNavigationBar = exports.setKeepScreenOn = exports.setClipboard = exports.setCanPullDown = exports.setBackgroundTextStyle = exports.setBackgroundColor = exports.sendSocketMessage = exports.SDKVersion = exports.scan = exports.saveImage = exports.saveFile = exports.rsa = exports.request = exports.reportAnalytics = exports.removeTabBarBadge = exports.removeStorageSync = exports.removeStorage = exports.removeSavedFile = exports.reLaunch = exports.redirectTo = exports.readBLECharacteristicValue = exports.prompt = exports.previewImage = exports.pageScrollTo = exports.optionsSelect = exports.openVoucherList = exports.openVoucherDetail = exports.openTicketList = exports.openTicketDetail = exports.openSetting = exports.openMerchantVoucherList = exports.openMerchantTicketList = exports.openMerchantCardList = exports.openLocation = exports.openKBVoucherDetail = exports.openCardList = exports.openCardDetail = exports.openBluetoothAdapter = exports.onUserCaptureScreen = exports.onSocketOpen = exports.onSocketMessage = exports.onSocketError = exports.onSocketClose = exports.onNetworkStatusChange = exports.onMemoryWarning = exports.onGyroscopeChange = void 0;
exports.httpRequest = exports.getOpenUserInfo = exports.createVideoContext = exports.writeBLECharacteristicValue = exports.watchShake = exports.vibrateShort = exports.vibrateLong = exports.vibrate = exports.uploadFile = exports.tradePay = exports.textRiskIdentification = exports.switchTab = exports.stopPullDownRefresh = exports.stopBluetoothDevicesDiscovery = exports.stopBeaconDiscovery = exports.startZMVerify = exports.startPullDownRefresh = exports.startBluetoothDevicesDiscovery = exports.startBeaconDiscovery = exports.showToast = exports.showTabBarRedDot = exports.showTabBar = exports.showSharePanel = exports.showNavigationBarLoading = exports.showLoading = exports.showAuthGuide = exports.showActionSheet = exports.setTabBarStyle = exports.setTabBarItem = exports.setTabBarBadge = void 0;
var framework_shared_1 = require("@remax/framework-shared");
exports.addCardAuth = framework_shared_1.promisify(my.addCardAuth);
exports.addPhoneContact = framework_shared_1.promisify(my.addPhoneContact);
exports.alert = framework_shared_1.promisify(my.alert);
exports.canIUse = my.canIUse;
exports.chooseAlipayContact = framework_shared_1.promisify(my.chooseAlipayContact);
exports.chooseCity = framework_shared_1.promisify(my.chooseCity);
exports.chooseContact = framework_shared_1.promisify(my.chooseContact);
exports.chooseImage = framework_shared_1.promisify(my.chooseImage);
exports.chooseLocation = framework_shared_1.promisify(my.chooseLocation);
exports.choosePhoneContact = framework_shared_1.promisify(my.choosePhoneContact);
exports.clearStorage = my.clearStorage;
exports.clearStorageSync = my.clearStorageSync;
exports.closeBluetoothAdapter = framework_shared_1.promisify(my.closeBluetoothAdapter);
exports.closeSocket = framework_shared_1.promisify(my.closeSocket);
exports.compressImage = framework_shared_1.promisify(my.compressImage);
exports.confirm = framework_shared_1.promisify(my.confirm);
exports.connectBLEDevice = framework_shared_1.promisify(my.connectBLEDevice);
exports.connectSocket = framework_shared_1.promisify(my.connectSocket);
exports.createAnimation = my.createAnimation;
exports.createCanvasContext = my.createCanvasContext;
exports.createIntersectionObserver = my.createIntersectionObserver;
exports.createMapContext = my.createMapContext;
exports.createSelectorQuery = my.createSelectorQuery;
exports.createWebViewContext = my.createWebViewContext;
exports.datePicker = framework_shared_1.promisify(my.datePicker);
exports.disconnectBLEDevice = framework_shared_1.promisify(my.disconnectBLEDevice);
exports.downloadFile = framework_shared_1.promisify(my.downloadFile);
exports.getAuthCode = framework_shared_1.promisify(my.getAuthCode);
exports.getAuthUserInfo = framework_shared_1.promisify(my.getAuthUserInfo);
exports.getBatteryInfo = framework_shared_1.promisify(my.getBatteryInfo);
exports.getBatteryInfoSync = my.getBatteryInfoSync;
exports.getBeacons = framework_shared_1.promisify(my.getBeacons);
exports.getBLEDeviceCharacteristics = framework_shared_1.promisify(my.getBLEDeviceCharacteristics);
exports.getBLEDeviceServices = framework_shared_1.promisify(my.getBLEDeviceServices);
exports.getBluetoothAdapterState = framework_shared_1.promisify(my.getBluetoothAdapterState);
exports.getBluetoothDevices = framework_shared_1.promisify(my.getBluetoothDevices);
exports.getClipboard = framework_shared_1.promisify(my.getClipboard);
exports.getConnectedBluetoothDevices = framework_shared_1.promisify(my.getConnectedBluetoothDevices);
exports.getFileInfo = framework_shared_1.promisify(my.getFileInfo);
exports.getImageInfo = framework_shared_1.promisify(my.getImageInfo);
exports.getLocation = framework_shared_1.promisify(my.getLocation);
exports.getNetworkType = framework_shared_1.promisify(my.getNetworkType);
exports.getPhoneNumber = framework_shared_1.promisify(my.getPhoneNumber);
exports.getRunData = framework_shared_1.promisify(my.getRunData);
exports.getRunScene = framework_shared_1.promisify(my.getRunScene);
exports.getSavedFileInfo = framework_shared_1.promisify(my.getSavedFileInfo);
exports.getSavedFileList = framework_shared_1.promisify(my.getSavedFileList);
exports.getScreenBrightness = framework_shared_1.promisify(my.getScreenBrightness);
exports.getServerTime = framework_shared_1.promisify(my.getServerTime);
exports.getSetting = framework_shared_1.promisify(my.getSetting);
exports.getStorage = framework_shared_1.promisify(my.getStorage);
exports.getStorageInfo = framework_shared_1.promisify(my.getStorageInfo);
exports.getStorageInfoSync = my.getStorageInfoSync;
exports.getStorageSync = my.getStorageSync;
exports.getSystemInfo = framework_shared_1.promisify(my.getSystemInfo);
exports.getSystemInfoSync = my.getSystemInfoSync;
exports.getTitleColor = framework_shared_1.promisify(my.getTitleColor);
exports.getUpdateManager = my.getUpdateManager;
exports.hideAddToDesktopMenu = my.hideAddToDesktopMenu;
exports.hideAllAddToDesktopMenu = my.hideAllAddToDesktopMenu;
exports.hideAllFavoriteMenu = my.hideAllFavoriteMenu;
exports.hideBackHome = my.hideBackHome;
exports.hideFavoriteMenu = my.hideFavoriteMenu;
exports.hideKeyboard = my.hideKeyboard;
exports.hideLoading = my.hideLoading;
exports.hideNavigationBarLoading = my.hideNavigationBarLoading;
exports.hideShareMenu = my.hideShareMenu;
exports.hideTabBar = my.hideTabBar;
exports.hideTabBarRedDot = my.hideTabBarRedDot;
exports.hideToast = my.hideToast;
exports.loadFontFace = framework_shared_1.promisify(my.loadFontFace);
exports.makePhoneCall = my.makePhoneCall;
exports.multiLevelSelect = framework_shared_1.promisify(my.multiLevelSelect);
exports.navigateBack = framework_shared_1.promisify(my.navigateBack);
exports.navigateBackMiniProgram = framework_shared_1.promisify(my.navigateBackMiniProgram);
exports.navigateTo = framework_shared_1.promisify(my.navigateTo);
exports.navigateToMiniProgram = framework_shared_1.promisify(my.navigateToMiniProgram);
exports.notifyBLECharacteristicValueChange = framework_shared_1.promisify(my.notifyBLECharacteristicValueChange);
exports.offAccelerometerChange = my.offAccelerometerChange;
exports.offBLECharacteristicValueChange = my.offBLECharacteristicValueChange;
exports.offBLEConnectionStateChanged = my.offBLEConnectionStateChanged;
exports.offBluetoothAdapterStateChange = my.offBluetoothAdapterStateChange;
exports.offBluetoothDeviceFound = my.offBluetoothDeviceFound;
exports.offCompassChange = my.offCompassChange;
exports.offGyroscopeChange = my.offGyroscopeChange;
exports.offMemoryWarning = my.offMemoryWarning;
exports.offNetworkStatusChange = my.offNetworkStatusChange;
exports.offSocketClose = my.offSocketClose;
exports.offSocketError = my.offSocketError;
exports.offSocketMessage = my.offSocketMessage;
exports.offSocketOpen = my.offSocketOpen;
exports.offUserCaptureScreen = my.offUserCaptureScreen;
exports.onAccelerometerChange = my.onAccelerometerChange;
exports.onBeaconServiceChange = my.onBeaconServiceChange;
exports.onBeaconUpdate = my.onBeaconUpdate;
exports.onBLECharacteristicValueChange = my.onBLECharacteristicValueChange;
exports.onBLEConnectionStateChanged = my.onBLEConnectionStateChanged;
exports.onBluetoothAdapterStateChange = my.onBluetoothAdapterStateChange;
exports.onBluetoothDeviceFound = my.onBluetoothDeviceFound;
exports.onCompassChange = my.onCompassChange;
exports.onGyroscopeChange = my.onGyroscopeChange;
exports.onMemoryWarning = my.onMemoryWarning;
exports.onNetworkStatusChange = my.onNetworkStatusChange;
exports.onSocketClose = my.onSocketClose;
exports.onSocketError = my.onSocketError;
exports.onSocketMessage = my.onSocketMessage;
exports.onSocketOpen = my.onSocketOpen;
exports.onUserCaptureScreen = my.onUserCaptureScreen;
exports.openBluetoothAdapter = my.openBluetoothAdapter;
exports.openCardDetail = my.openCardDetail;
exports.openCardList = my.openCardList;
exports.openKBVoucherDetail = my.openKBVoucherDetail;
exports.openLocation = framework_shared_1.promisify(my.openLocation);
exports.openMerchantCardList = my.openMerchantCardList;
exports.openMerchantTicketList = my.openMerchantTicketList;
exports.openMerchantVoucherList = my.openMerchantVoucherList;
exports.openSetting = framework_shared_1.promisify(my.openSetting);
exports.openTicketDetail = my.openTicketDetail;
exports.openTicketList = my.openTicketList;
exports.openVoucherDetail = my.openVoucherDetail;
exports.openVoucherList = my.openVoucherList;
exports.optionsSelect = my.optionsSelect;
exports.pageScrollTo = framework_shared_1.promisify(my.pageScrollTo);
exports.previewImage = framework_shared_1.promisify(my.previewImage);
exports.prompt = framework_shared_1.promisify(my.prompt);
exports.readBLECharacteristicValue = framework_shared_1.promisify(my.readBLECharacteristicValue);
exports.redirectTo = framework_shared_1.promisify(my.redirectTo);
exports.reLaunch = framework_shared_1.promisify(my.reLaunch);
exports.removeSavedFile = framework_shared_1.promisify(my.removeSavedFile);
exports.removeStorage = framework_shared_1.promisify(my.removeStorage);
exports.removeStorageSync = my.removeStorageSync;
exports.removeTabBarBadge = framework_shared_1.promisify(my.removeTabBarBadge);
exports.reportAnalytics = my.reportAnalytics;
exports.request = framework_shared_1.promisify(my.request);
exports.rsa = framework_shared_1.promisify(my.rsa);
exports.saveFile = framework_shared_1.promisify(my.saveFile);
exports.saveImage = framework_shared_1.promisify(my.saveImage);
exports.scan = framework_shared_1.promisify(my.scan);
exports.SDKVersion = my.SDKVersion;
exports.sendSocketMessage = framework_shared_1.promisify(my.sendSocketMessage);
exports.setBackgroundColor = framework_shared_1.promisify(my.setBackgroundColor);
exports.setBackgroundTextStyle = framework_shared_1.promisify(my.setBackgroundTextStyle);
exports.setCanPullDown = my.setCanPullDown;
exports.setClipboard = framework_shared_1.promisify(my.setClipboard);
exports.setKeepScreenOn = framework_shared_1.promisify(my.setKeepScreenOn);
exports.setNavigationBar = framework_shared_1.promisify(my.setNavigationBar);
exports.setOptionMenu = my.setOptionMenu;
exports.setScreenBrightness = framework_shared_1.promisify(my.setScreenBrightness);
exports.setStorage = framework_shared_1.promisify(my.setStorage);
exports.setStorageSync = my.setStorageSync;
exports.setTabBarBadge = framework_shared_1.promisify(my.setTabBarBadge);
exports.setTabBarItem = framework_shared_1.promisify(my.setTabBarItem);
exports.setTabBarStyle = framework_shared_1.promisify(my.setTabBarStyle);
exports.showActionSheet = framework_shared_1.promisify(my.showActionSheet);
exports.showAuthGuide = my.showAuthGuide;
exports.showLoading = framework_shared_1.promisify(my.showLoading);
exports.showNavigationBarLoading = framework_shared_1.promisify(my.showNavigationBarLoading);
exports.showSharePanel = my.showSharePanel;
exports.showTabBar = framework_shared_1.promisify(my.showTabBar);
exports.showTabBarRedDot = framework_shared_1.promisify(my.showTabBarRedDot);
exports.showToast = framework_shared_1.promisify(my.showToast);
exports.startBeaconDiscovery = framework_shared_1.promisify(my.startBeaconDiscovery);
exports.startBluetoothDevicesDiscovery = framework_shared_1.promisify(my.startBluetoothDevicesDiscovery);
exports.startPullDownRefresh = framework_shared_1.promisify(my.startPullDownRefresh);
exports.startZMVerify = my.startZMVerify;
exports.stopBeaconDiscovery = framework_shared_1.promisify(my.stopBeaconDiscovery);
exports.stopBluetoothDevicesDiscovery = framework_shared_1.promisify(my.stopBluetoothDevicesDiscovery);
exports.stopPullDownRefresh = framework_shared_1.promisify(my.stopPullDownRefresh);
exports.switchTab = framework_shared_1.promisify(my.switchTab);
exports.textRiskIdentification = framework_shared_1.promisify(my.textRiskIdentification);
exports.tradePay = framework_shared_1.promisify(my.tradePay);
exports.uploadFile = framework_shared_1.promisify(my.uploadFile);
exports.vibrate = framework_shared_1.promisify(my.vibrate);
exports.vibrateLong = framework_shared_1.promisify(my.vibrateLong);
exports.vibrateShort = framework_shared_1.promisify(my.vibrateShort);
exports.watchShake = my.watchShake;
exports.writeBLECharacteristicValue = framework_shared_1.promisify(my.writeBLECharacteristicValue);
exports.createVideoContext = my.createVideoContext;
exports.getOpenUserInfo = framework_shared_1.promisify(my.getOpenUserInfo);
exports.httpRequest = framework_shared_1.promisify(my.httpRequest);
