// @remax/cli/lib/build/WebBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const webpack_dev_server_1 = __importDefault(require("webpack-dev-server"));
const detect_port_1 = __importDefault(require("detect-port"));
const config_web_1 = __importDefault(require("./webpack/config.web"));
const config_web_mpa_1 = __importDefault(require("./webpack/config.web.mpa"));
const address_1 = __importDefault(require("address"));
const output_1 = __importDefault(require("./utils/output"));
const watch_1 = __importDefault(require("./watch"));
const Builder_1 = __importDefault(require("./Builder"));
class WebBuilder extends Builder_1.default {
    constructor(api, options) {
        super(api, options, 'webapp');
    }
    createWebpackConfig() {
        var _a;
        return ((_a = this.options.web) === null || _a === void 0 ? void 0 : _a.mpa) ? config_web_mpa_1.default(this) : config_web_1.default(this);
    }
    run() {
        if (this.options.watch) {
            this.watch();
        }
        else {
            this.build();
        }
        return this.webpackCompiler;
    }
    watch() {
        var _a;
        const designatedPort = (_a = this.options.port) !== null && _a !== void 0 ? _a : 3000;
        detect_port_1.default(designatedPort, (err, port) => {
            if (err) {
                output_1.default.error(err.message);
                process.exit(1);
            }
            if (designatedPort !== port) {
                output_1.default.warn(` ç«¯å£: ${designatedPort} è¢«å ç”¨ï¼Œç³»ç»Ÿå·²åˆ†é…å¦ä¸€ä¸ªå¯ç”¨ç«¯å£ï¼š${port}`);
            }
            output_1.default.message('ðŸš€ å¯åŠ¨ watch', 'blue');
            output_1.default.message(`ðŸ“Ž http://localhost:${port}`, 'blue');
            output_1.default.message(`ðŸ“Ž http://${address_1.default.ip()}:${port}\n`, 'blue');
            const server = new webpack_dev_server_1.default(this.webpackCompiler, this.webpackConfig.devServer);
            this.webpackCompiler.hooks.done.tap('web-dev', stats => {
                console.log(stats.toString({
                    colors: true,
                    modules: false,
                    children: false,
                    assets: false,
                    entrypoints: false,
                }));
            });
            server.listen(port, '0.0.0.0', error => {
                if (error) {
                    console.error(error);
                    process.exit(1);
                }
            });
            watch_1.default(this, server);
        });
    }
    build() {
        this.webpackCompiler.run((error, stats) => {
            if (error) {
                output_1.default.error(error.message);
                throw error;
            }
            const info = stats.toJson();
            if (stats.hasErrors()) {
                info.errors.forEach(error => {
                    output_1.default.error(error);
                });
                process.exit(1);
            }
            if (stats.hasWarnings()) {
                info.warnings.forEach(warning => {
                    output_1.default.warn(warning);
                });
            }
        });
    }
}
exports.default = WebBuilder;
