// @remax/wechat/cjs/hostComponents/LivePusher/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.props = exports.alias = void 0;
exports.alias = {
    id: 'id',
    className: 'class',
    style: 'style',
    animation: 'animation',
    url: 'url',
    mode: 'mode',
    autopush: 'autopush',
    muted: 'muted',
    enableCamera: 'enable-camera',
    autoFocus: 'auto-focus',
    orientation: 'orientation',
    beauty: 'beauty',
    whiteness: 'whiteness',
    aspect: 'aspect',
    minBitrate: 'min-bitrate',
    maxBitrate: 'max-bitrate',
    audioQuality: 'audio-quality',
    waitingImage: 'waiting-image',
    waitingImageHash: 'waiting-image-hash',
    zoom: 'zoom',
    devicePosition: 'device-position',
    backgroundMute: 'background-mute',
    mirror: 'mirror',
    remoteMirror: 'remote-mirror',
    localMirror: 'local-mirror',
    audioReverbType: 'audio-reverb-type',
    enableMic: 'enable-mic',
    enableAgc: 'enable-agc',
    enableAns: 'enable-ans',
    audioVolumeType: 'audio-volume-type',
    videoWidth: 'video-width',
    videoHeight: 'video-height',
    onStateChange: 'bindstatechange',
    onNetStatus: 'bindnetstatus',
    onError: 'binderror',
    onBgmStart: 'bindbgmstart',
    onBgmProgress: 'bindbgmprogress',
    onBgmComplete: 'bindbgmcomplete',
};
exports.props = Object.values(exports.alias);
