// @remax/wechat/esm/api/index.js
import { promisify } from '@remax/framework-shared';
export var canIUse = wx.canIUse;
export var base64ToArrayBuffer = wx.base64ToArrayBuffer;
export var arrayBufferToBase64 = wx.arrayBufferToBase64;
export var getSystemInfoSync = wx.getSystemInfoSync;
export var getSystemInfo = promisify(wx.getSystemInfo);
export var getUpdateManager = wx.getUpdateManager;
export var getLaunchOptionsSync = wx.getLaunchOptionsSync;
export var onPageNotFound = wx.onPageNotFound;
export var onError = wx.onError;
export var onAudioInterruptionEnd = wx.onAudioInterruptionEnd;
export var onAppShow = wx.onAppShow;
export var onAppHide = wx.onAppHide;
export var offPageNotFound = wx.offPageNotFound;
export var offError = wx.offError;
export var offAudioInterruptionEnd = wx.offAudioInterruptionEnd;
export var offAudioInterruptionBegin = wx.offAudioInterruptionBegin;
export var offAppShow = wx.offAppShow;
export var offAppHide = wx.offAppHide;
export var setEnableDebug = promisify(wx.setEnableDebug);
export var getLogManager = wx.getLogManager;
export var switchTab = promisify(wx.switchTab);
export var reLaunch = promisify(wx.reLaunch);
export var redirectTo = promisify(wx.redirectTo);
export var navigateTo = promisify(wx.navigateTo);
export var navigateBack = promisify(wx.navigateBack);
export var showToast = promisify(wx.showToast);
export var showModal = promisify(wx.showModal);
export var showLoading = promisify(wx.showLoading);
export var showActionSheet = promisify(wx.showActionSheet);
export var hideToast = promisify(wx.hideToast);
export var hideLoading = promisify(wx.hideLoading);
export var showNavigationBarLoading = promisify(wx.showNavigationBarLoading);
export var setNavigationBarTitle = promisify(wx.setNavigationBarTitle);
export var setNavigationBarColor = promisify(wx.setNavigationBarColor);
export var hideNavigationBarLoading = promisify(wx.hideNavigationBarLoading);
export var setBackgroundTextStyle = promisify(wx.setBackgroundTextStyle);
export var setBackgroundColor = promisify(wx.setBackgroundColor);
export var showTabBarRedDot = promisify(wx.showTabBarRedDot);
export var showTabBar = promisify(wx.showTabBar);
export var setTabBarStyle = promisify(wx.setTabBarStyle);
export var setTabBarItem = promisify(wx.setTabBarItem);
export var setTabBarBadge = promisify(wx.setTabBarBadge);
export var removeTabBarBadge = promisify(wx.removeTabBarBadge);
export var hideTabBarRedDot = promisify(wx.hideTabBarRedDot);
export var hideTabBar = promisify(wx.hideTabBar);
export var loadFontFace = promisify(wx.loadFontFace);
export var stopPullDownRefresh = promisify(wx.stopPullDownRefresh);
export var startPullDownRefresh = promisify(wx.startPullDownRefresh);
export var pageScrollTo = promisify(wx.pageScrollTo);
export var createAnimation = wx.createAnimation;
export var setTopBarText = promisify(wx.setTopBarText);
export var nextTick = wx.nextTick;
export var getMenuButtonBoundingClientRect = wx.getMenuButtonBoundingClientRect;
export var onWindowResize = wx.onWindowResize;
export var offWindowResize = wx.offWindowResize;
export var onKeyboardHeightChange = wx.onKeyboardHeightChange;
export var offKeyboardHeightChange = wx.offKeyboardHeightChange;
export var getSelectedTextRange = wx.getSelectedTextRange;
export var request = promisify(wx.request);
export var downloadFile = promisify(wx.downloadFile);
export var uploadFile = promisify(wx.uploadFile);
export var sendSocketMessage = promisify(wx.sendSocketMessage);
export var onSocketOpen = wx.onSocketOpen;
export var onSocketMessage = wx.onSocketMessage;
export var onSocketError = wx.onSocketError;
export var onSocketClose = wx.onSocketClose;
export var connectSocket = wx.connectSocket;
export var closeSocket = promisify(wx.closeSocket);
export var stopLocalServiceDiscovery = promisify(wx.stopLocalServiceDiscovery);
export var startLocalServiceDiscovery = promisify(wx.startLocalServiceDiscovery);
export var onLocalServiceResolveFail = wx.onLocalServiceResolveFail;
export var onLocalServiceLost = wx.onLocalServiceLost;
export var onLocalServiceFound = wx.onLocalServiceFound;
export var onLocalServiceDiscoveryStop = wx.onLocalServiceDiscoveryStop;
export var offLocalServiceResolveFail = wx.offLocalServiceResolveFail;
export var offLocalServiceLost = wx.offLocalServiceLost;
export var offLocalServiceFound = wx.offLocalServiceFound;
export var offLocalServiceDiscoveryStop = wx.offLocalServiceDiscoveryStop;
export var createUDPSocket = wx.createUDPSocket;
export var setStorageSync = wx.setStorageSync;
export var setStorage = promisify(wx.setStorage);
export var removeStorageSync = wx.removeStorageSync;
export var removeStorage = promisify(wx.removeStorage);
export var getStorageSync = wx.getStorageSync;
export var getStorageInfoSync = wx.getStorageInfoSync;
export var getStorageInfo = promisify(wx.getStorageInfo);
export var getStorage = promisify(wx.getStorage);
export var clearStorageSync = wx.clearStorageSync;
export var clearStorage = promisify(wx.clearStorage);
export var createMapContext = wx.createMapContext;
export var saveImageToPhotosAlbum = promisify(wx.saveImageToPhotosAlbum);
export var previewImage = promisify(wx.previewImage);
export var getImageInfo = promisify(wx.getImageInfo);
export var compressImage = promisify(wx.compressImage);
export var chooseMessageFile = promisify(wx.chooseMessageFile);
export var chooseImage = promisify(wx.chooseImage);
export var saveVideoToPhotosAlbum = promisify(wx.saveVideoToPhotosAlbum);
export var createVideoContext = wx.createVideoContext;
export var chooseVideo = promisify(wx.chooseVideo);
export var chooseMedia = promisify(wx.chooseMedia);
export var stopVoice = promisify(wx.stopVoice);
export var setInnerAudioOption = promisify(wx.setInnerAudioOption);
export var playVoice = promisify(wx.playVoice);
export var pauseVoice = promisify(wx.pauseVoice);
export var getAvailableAudioSources = promisify(wx.getAvailableAudioSources);
export var createInnerAudioContext = wx.createInnerAudioContext;
export var createAudioContext = wx.createAudioContext;
export var stopBackgroundAudio = promisify(wx.stopBackgroundAudio);
export var seekBackgroundAudio = promisify(wx.seekBackgroundAudio);
export var playBackgroundAudio = promisify(wx.playBackgroundAudio);
export var pauseBackgroundAudio = promisify(wx.pauseBackgroundAudio);
export var onBackgroundAudioStop = wx.onBackgroundAudioStop;
export var onBackgroundAudioPlay = wx.onBackgroundAudioPlay;
export var onBackgroundAudioPause = wx.onBackgroundAudioPause;
export var getBackgroundAudioPlayerState = promisify(wx.getBackgroundAudioPlayerState);
export var getBackgroundAudioManager = wx.getBackgroundAudioManager;
export var createLivePusherContext = wx.createLivePusherContext;
export var createLivePlayerContext = wx.createLivePlayerContext;
export var stopRecord = promisify(wx.stopRecord);
export var startRecord = promisify(wx.startRecord);
export var getRecorderManager = wx.getRecorderManager;
export var createCameraContext = wx.createCameraContext;
export var stopLocationUpdate = promisify(wx.stopLocationUpdate);
export var startLocationUpdateBackground = promisify(wx.startLocationUpdateBackground);
export var startLocationUpdate = promisify(wx.startLocationUpdate);
export var openLocation = promisify(wx.openLocation);
export var onLocationChange = wx.onLocationChange;
export var getLocation = promisify(wx.getLocation);
export var chooseLocation = promisify(wx.chooseLocation);
export var updateShareMenu = promisify(wx.updateShareMenu);
export var showShareMenu = promisify(wx.showShareMenu);
export var hideShareMenu = promisify(wx.hideShareMenu);
export var getShareInfo = promisify(wx.getShareInfo);
export var createOffscreenCanvas = wx.createOffscreenCanvas;
export var createCanvasContext = wx.createCanvasContext;
export var canvasToTempFilePath = promisify(wx.canvasToTempFilePath);
export var canvasPutImageData = promisify(wx.canvasPutImageData);
export var canvasGetImageData = promisify(wx.canvasGetImageData);
export var saveFile = promisify(wx.saveFile);
export var removeSavedFile = promisify(wx.removeSavedFile);
export var openDocument = promisify(wx.openDocument);
export var getSavedFileList = promisify(wx.getSavedFileList);
export var getSavedFileInfo = promisify(wx.getSavedFileInfo);
export var getFileSystemManager = wx.getFileSystemManager;
export var getFileInfo = promisify(wx.getFileInfo);
export var login = promisify(wx.login);
export var checkSession = promisify(wx.checkSession);
export var navigateToMiniProgram = promisify(wx.navigateToMiniProgram);
export var navigateBackMiniProgram = promisify(wx.navigateBackMiniProgram);
export var getAccountInfoSync = wx.getAccountInfoSync;
export var getUserInfo = promisify(wx.getUserInfo);
export var reportMonitor = wx.reportMonitor;
export var reportAnalytics = wx.reportAnalytics;
export var requestPayment = promisify(wx.requestPayment);
export var authorize = promisify(wx.authorize);
export var openSetting = promisify(wx.openSetting);
export var getSetting = promisify(wx.getSetting);
export var chooseAddress = promisify(wx.chooseAddress);
export var openCard = promisify(wx.openCard);
export var addCard = promisify(wx.addCard);
export var chooseInvoiceTitle = promisify(wx.chooseInvoiceTitle);
export var chooseInvoice = promisify(wx.chooseInvoice);
export var startSoterAuthentication = promisify(wx.startSoterAuthentication);
export var checkIsSupportSoterAuthentication = promisify(wx.checkIsSupportSoterAuthentication);
export var checkIsSoterEnrolledInDevice = promisify(wx.checkIsSoterEnrolledInDevice);
export var getWeRunData = promisify(wx.getWeRunData);
export var stopBeaconDiscovery = promisify(wx.stopBeaconDiscovery);
export var startBeaconDiscovery = promisify(wx.startBeaconDiscovery);
export var onBeaconUpdate = wx.onBeaconUpdate;
export var offBeaconUpdate = wx.offBeaconUpdate;
export var onBeaconServiceChange = wx.onBeaconServiceChange;
export var offBeaconServiceChange = wx.offBeaconServiceChange;
export var getBeacons = promisify(wx.getBeacons);
export var stopWifi = promisify(wx.stopWifi);
export var startWifi = promisify(wx.startWifi);
export var setWifiList = promisify(wx.setWifiList);
export var onWifiConnected = wx.onWifiConnected;
export var offWifiConnected = wx.offWifiConnected;
export var onGetWifiList = wx.onGetWifiList;
export var offGetWifiList = wx.offGetWifiList;
export var getWifiList = promisify(wx.getWifiList);
export var getConnectedWifi = promisify(wx.getConnectedWifi);
export var connectWifi = promisify(wx.connectWifi);
export var addPhoneContact = promisify(wx.addPhoneContact);
export var writeBLECharacteristicValue = promisify(wx.writeBLECharacteristicValue);
export var readBLECharacteristicValue = promisify(wx.readBLECharacteristicValue);
export var onBLEConnectionStateChange = wx.onBLEConnectionStateChange;
export var offBLEConnectionStateChange = wx.offBLEConnectionStateChange;
export var onBLECharacteristicValueChange = wx.onBLECharacteristicValueChange;
export var offBLECharacteristicValueChange = wx.offBLECharacteristicValueChange;
export var notifyBLECharacteristicValueChange = promisify(wx.notifyBLECharacteristicValueChange);
export var getBLEDeviceServices = promisify(wx.getBLEDeviceServices);
export var getBLEDeviceCharacteristics = promisify(wx.getBLEDeviceCharacteristics);
export var createBLEConnection = promisify(wx.createBLEConnection);
export var closeBLEConnection = promisify(wx.closeBLEConnection);
export var stopBluetoothDevicesDiscovery = promisify(wx.stopBluetoothDevicesDiscovery);
export var startBluetoothDevicesDiscovery = promisify(wx.startBluetoothDevicesDiscovery);
export var openBluetoothAdapter = wx.openBluetoothAdapter;
export var onBluetoothDeviceFound = wx.onBluetoothDeviceFound;
export var offBluetoothDeviceFound = wx.offBluetoothDeviceFound;
export var onBluetoothAdapterStateChange = wx.onBluetoothAdapterStateChange;
export var offBluetoothAdapterStateChange = wx.offBluetoothAdapterStateChange;
export var getConnectedBluetoothDevices = promisify(wx.getConnectedBluetoothDevices);
export var getBluetoothDevices = promisify(wx.getBluetoothDevices);
export var getBluetoothAdapterState = promisify(wx.getBluetoothAdapterState);
export var closeBluetoothAdapter = promisify(wx.closeBluetoothAdapter);
export var getBatteryInfoSync = wx.getBatteryInfoSync;
export var getBatteryInfo = promisify(wx.getBatteryInfo);
export var setClipboardData = promisify(wx.setClipboardData);
export var getClipboardData = promisify(wx.getClipboardData);
export var stopHCE = promisify(wx.stopHCE);
export var startHCE = promisify(wx.startHCE);
export var sendHCEMessage = promisify(wx.sendHCEMessage);
export var onHCEMessage = wx.onHCEMessage;
export var offHCEMessage = wx.offHCEMessage;
export var getHCEState = promisify(wx.getHCEState);
export var onNetworkStatusChange = wx.onNetworkStatusChange;
export var offNetworkStatusChange = wx.offNetworkStatusChange;
export var getNetworkType = promisify(wx.getNetworkType);
export var setScreenBrightness = promisify(wx.setScreenBrightness);
export var setKeepScreenOn = promisify(wx.setKeepScreenOn);
export var onUserCaptureScreen = wx.onUserCaptureScreen;
export var offUserCaptureScreen = wx.offUserCaptureScreen;
export var getScreenBrightness = promisify(wx.getScreenBrightness);
export var makePhoneCall = promisify(wx.makePhoneCall);
export var stopAccelerometer = promisify(wx.stopAccelerometer);
export var startAccelerometer = promisify(wx.startAccelerometer);
export var onAccelerometerChange = wx.onAccelerometerChange;
export var offAccelerometerChange = wx.offAccelerometerChange;
export var stopCompass = promisify(wx.stopCompass);
export var startCompass = promisify(wx.startCompass);
export var onCompassChange = wx.onCompassChange;
export var offCompassChange = wx.offCompassChange;
export var stopDeviceMotionListening = promisify(wx.stopDeviceMotionListening);
export var startDeviceMotionListening = promisify(wx.startDeviceMotionListening);
export var onDeviceMotionChange = wx.onDeviceMotionChange;
export var offDeviceMotionChange = wx.offDeviceMotionChange;
export var stopGyroscope = promisify(wx.stopGyroscope);
export var startGyroscope = promisify(wx.startGyroscope);
export var onGyroscopeChange = wx.onGyroscopeChange;
export var offGyroscopeChange = wx.offGyroscopeChange;
export var onMemoryWarning = wx.onMemoryWarning;
export var offMemoryWarning = wx.offMemoryWarning;
export var scanCode = promisify(wx.scanCode);
export var vibrateShort = promisify(wx.vibrateShort);
export var vibrateLong = promisify(wx.vibrateLong);
export var createWorker = wx.createWorker;
export var getExtConfigSync = wx.getExtConfigSync;
export var getExtConfig = promisify(wx.getExtConfig);
export var createSelectorQuery = wx.createSelectorQuery;
export var createIntersectionObserver = wx.createIntersectionObserver;
export var createRewardedVideoAd = wx.createRewardedVideoAd;
export var createInterstitialAd = wx.createInterstitialAd;
export var cloud = wx.cloud;
export var requestSubscribeMessage = promisify(wx.requestSubscribeMessage);
export var hideHomeButton = promisify(wx.hideHomeButton);
export var getRealtimeLogManager = wx.getRealtimeLogManager;
export var hideKeyboard = promisify(wx.hideKeyboard);
