// @remax/wechat/cjs/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAnimation = exports.pageScrollTo = exports.startPullDownRefresh = exports.stopPullDownRefresh = exports.loadFontFace = exports.hideTabBar = exports.hideTabBarRedDot = exports.removeTabBarBadge = exports.setTabBarBadge = exports.setTabBarItem = exports.setTabBarStyle = exports.showTabBar = exports.showTabBarRedDot = exports.setBackgroundColor = exports.setBackgroundTextStyle = exports.hideNavigationBarLoading = exports.setNavigationBarColor = exports.setNavigationBarTitle = exports.showNavigationBarLoading = exports.hideLoading = exports.hideToast = exports.showActionSheet = exports.showLoading = exports.showModal = exports.showToast = exports.navigateBack = exports.navigateTo = exports.redirectTo = exports.reLaunch = exports.switchTab = exports.getLogManager = exports.setEnableDebug = exports.offAppHide = exports.offAppShow = exports.offAudioInterruptionBegin = exports.offAudioInterruptionEnd = exports.offError = exports.offPageNotFound = exports.onAppHide = exports.onAppShow = exports.onAudioInterruptionEnd = exports.onError = exports.onPageNotFound = exports.getLaunchOptionsSync = exports.getUpdateManager = exports.getSystemInfo = exports.getSystemInfoSync = exports.arrayBufferToBase64 = exports.base64ToArrayBuffer = exports.canIUse = void 0;
exports.chooseMedia = exports.chooseVideo = exports.createVideoContext = exports.saveVideoToPhotosAlbum = exports.chooseImage = exports.chooseMessageFile = exports.compressImage = exports.getImageInfo = exports.previewImage = exports.saveImageToPhotosAlbum = exports.createMapContext = exports.clearStorage = exports.clearStorageSync = exports.getStorage = exports.getStorageInfo = exports.getStorageInfoSync = exports.getStorageSync = exports.removeStorage = exports.removeStorageSync = exports.setStorage = exports.setStorageSync = exports.createUDPSocket = exports.offLocalServiceDiscoveryStop = exports.offLocalServiceFound = exports.offLocalServiceLost = exports.offLocalServiceResolveFail = exports.onLocalServiceDiscoveryStop = exports.onLocalServiceFound = exports.onLocalServiceLost = exports.onLocalServiceResolveFail = exports.startLocalServiceDiscovery = exports.stopLocalServiceDiscovery = exports.closeSocket = exports.connectSocket = exports.onSocketClose = exports.onSocketError = exports.onSocketMessage = exports.onSocketOpen = exports.sendSocketMessage = exports.uploadFile = exports.downloadFile = exports.request = exports.getSelectedTextRange = exports.offKeyboardHeightChange = exports.onKeyboardHeightChange = exports.offWindowResize = exports.onWindowResize = exports.getMenuButtonBoundingClientRect = exports.nextTick = exports.setTopBarText = void 0;
exports.getAccountInfoSync = exports.navigateBackMiniProgram = exports.navigateToMiniProgram = exports.checkSession = exports.login = exports.getFileInfo = exports.getFileSystemManager = exports.getSavedFileInfo = exports.getSavedFileList = exports.openDocument = exports.removeSavedFile = exports.saveFile = exports.canvasGetImageData = exports.canvasPutImageData = exports.canvasToTempFilePath = exports.createCanvasContext = exports.createOffscreenCanvas = exports.getShareInfo = exports.hideShareMenu = exports.showShareMenu = exports.updateShareMenu = exports.chooseLocation = exports.getLocation = exports.onLocationChange = exports.openLocation = exports.startLocationUpdate = exports.startLocationUpdateBackground = exports.stopLocationUpdate = exports.createCameraContext = exports.getRecorderManager = exports.startRecord = exports.stopRecord = exports.createLivePlayerContext = exports.createLivePusherContext = exports.getBackgroundAudioManager = exports.getBackgroundAudioPlayerState = exports.onBackgroundAudioPause = exports.onBackgroundAudioPlay = exports.onBackgroundAudioStop = exports.pauseBackgroundAudio = exports.playBackgroundAudio = exports.seekBackgroundAudio = exports.stopBackgroundAudio = exports.createAudioContext = exports.createInnerAudioContext = exports.getAvailableAudioSources = exports.pauseVoice = exports.playVoice = exports.setInnerAudioOption = exports.stopVoice = void 0;
exports.offBluetoothDeviceFound = exports.onBluetoothDeviceFound = exports.openBluetoothAdapter = exports.startBluetoothDevicesDiscovery = exports.stopBluetoothDevicesDiscovery = exports.closeBLEConnection = exports.createBLEConnection = exports.getBLEDeviceCharacteristics = exports.getBLEDeviceServices = exports.notifyBLECharacteristicValueChange = exports.offBLECharacteristicValueChange = exports.onBLECharacteristicValueChange = exports.offBLEConnectionStateChange = exports.onBLEConnectionStateChange = exports.readBLECharacteristicValue = exports.writeBLECharacteristicValue = exports.addPhoneContact = exports.connectWifi = exports.getConnectedWifi = exports.getWifiList = exports.offGetWifiList = exports.onGetWifiList = exports.offWifiConnected = exports.onWifiConnected = exports.setWifiList = exports.startWifi = exports.stopWifi = exports.getBeacons = exports.offBeaconServiceChange = exports.onBeaconServiceChange = exports.offBeaconUpdate = exports.onBeaconUpdate = exports.startBeaconDiscovery = exports.stopBeaconDiscovery = exports.getWeRunData = exports.checkIsSoterEnrolledInDevice = exports.checkIsSupportSoterAuthentication = exports.startSoterAuthentication = exports.chooseInvoice = exports.chooseInvoiceTitle = exports.addCard = exports.openCard = exports.chooseAddress = exports.getSetting = exports.openSetting = exports.authorize = exports.requestPayment = exports.reportAnalytics = exports.reportMonitor = exports.getUserInfo = void 0;
exports.createSelectorQuery = exports.getExtConfig = exports.getExtConfigSync = exports.createWorker = exports.vibrateLong = exports.vibrateShort = exports.scanCode = exports.offMemoryWarning = exports.onMemoryWarning = exports.offGyroscopeChange = exports.onGyroscopeChange = exports.startGyroscope = exports.stopGyroscope = exports.offDeviceMotionChange = exports.onDeviceMotionChange = exports.startDeviceMotionListening = exports.stopDeviceMotionListening = exports.offCompassChange = exports.onCompassChange = exports.startCompass = exports.stopCompass = exports.offAccelerometerChange = exports.onAccelerometerChange = exports.startAccelerometer = exports.stopAccelerometer = exports.makePhoneCall = exports.getScreenBrightness = exports.offUserCaptureScreen = exports.onUserCaptureScreen = exports.setKeepScreenOn = exports.setScreenBrightness = exports.getNetworkType = exports.offNetworkStatusChange = exports.onNetworkStatusChange = exports.getHCEState = exports.offHCEMessage = exports.onHCEMessage = exports.sendHCEMessage = exports.startHCE = exports.stopHCE = exports.getClipboardData = exports.setClipboardData = exports.getBatteryInfo = exports.getBatteryInfoSync = exports.closeBluetoothAdapter = exports.getBluetoothAdapterState = exports.getBluetoothDevices = exports.getConnectedBluetoothDevices = exports.offBluetoothAdapterStateChange = exports.onBluetoothAdapterStateChange = void 0;
exports.hideKeyboard = exports.getRealtimeLogManager = exports.hideHomeButton = exports.requestSubscribeMessage = exports.cloud = exports.createInterstitialAd = exports.createRewardedVideoAd = exports.createIntersectionObserver = void 0;
var framework_shared_1 = require("@remax/framework-shared");
exports.canIUse = wx.canIUse;
exports.base64ToArrayBuffer = wx.base64ToArrayBuffer;
exports.arrayBufferToBase64 = wx.arrayBufferToBase64;
exports.getSystemInfoSync = wx.getSystemInfoSync;
exports.getSystemInfo = framework_shared_1.promisify(wx.getSystemInfo);
exports.getUpdateManager = wx.getUpdateManager;
exports.getLaunchOptionsSync = wx.getLaunchOptionsSync;
exports.onPageNotFound = wx.onPageNotFound;
exports.onError = wx.onError;
exports.onAudioInterruptionEnd = wx.onAudioInterruptionEnd;
exports.onAppShow = wx.onAppShow;
exports.onAppHide = wx.onAppHide;
exports.offPageNotFound = wx.offPageNotFound;
exports.offError = wx.offError;
exports.offAudioInterruptionEnd = wx.offAudioInterruptionEnd;
exports.offAudioInterruptionBegin = wx.offAudioInterruptionBegin;
exports.offAppShow = wx.offAppShow;
exports.offAppHide = wx.offAppHide;
exports.setEnableDebug = framework_shared_1.promisify(wx.setEnableDebug);
exports.getLogManager = wx.getLogManager;
exports.switchTab = framework_shared_1.promisify(wx.switchTab);
exports.reLaunch = framework_shared_1.promisify(wx.reLaunch);
exports.redirectTo = framework_shared_1.promisify(wx.redirectTo);
exports.navigateTo = framework_shared_1.promisify(wx.navigateTo);
exports.navigateBack = framework_shared_1.promisify(wx.navigateBack);
exports.showToast = framework_shared_1.promisify(wx.showToast);
exports.showModal = framework_shared_1.promisify(wx.showModal);
exports.showLoading = framework_shared_1.promisify(wx.showLoading);
exports.showActionSheet = framework_shared_1.promisify(wx.showActionSheet);
exports.hideToast = framework_shared_1.promisify(wx.hideToast);
exports.hideLoading = framework_shared_1.promisify(wx.hideLoading);
exports.showNavigationBarLoading = framework_shared_1.promisify(wx.showNavigationBarLoading);
exports.setNavigationBarTitle = framework_shared_1.promisify(wx.setNavigationBarTitle);
exports.setNavigationBarColor = framework_shared_1.promisify(wx.setNavigationBarColor);
exports.hideNavigationBarLoading = framework_shared_1.promisify(wx.hideNavigationBarLoading);
exports.setBackgroundTextStyle = framework_shared_1.promisify(wx.setBackgroundTextStyle);
exports.setBackgroundColor = framework_shared_1.promisify(wx.setBackgroundColor);
exports.showTabBarRedDot = framework_shared_1.promisify(wx.showTabBarRedDot);
exports.showTabBar = framework_shared_1.promisify(wx.showTabBar);
exports.setTabBarStyle = framework_shared_1.promisify(wx.setTabBarStyle);
exports.setTabBarItem = framework_shared_1.promisify(wx.setTabBarItem);
exports.setTabBarBadge = framework_shared_1.promisify(wx.setTabBarBadge);
exports.removeTabBarBadge = framework_shared_1.promisify(wx.removeTabBarBadge);
exports.hideTabBarRedDot = framework_shared_1.promisify(wx.hideTabBarRedDot);
exports.hideTabBar = framework_shared_1.promisify(wx.hideTabBar);
exports.loadFontFace = framework_shared_1.promisify(wx.loadFontFace);
exports.stopPullDownRefresh = framework_shared_1.promisify(wx.stopPullDownRefresh);
exports.startPullDownRefresh = framework_shared_1.promisify(wx.startPullDownRefresh);
exports.pageScrollTo = framework_shared_1.promisify(wx.pageScrollTo);
exports.createAnimation = wx.createAnimation;
exports.setTopBarText = framework_shared_1.promisify(wx.setTopBarText);
exports.nextTick = wx.nextTick;
exports.getMenuButtonBoundingClientRect = wx.getMenuButtonBoundingClientRect;
exports.onWindowResize = wx.onWindowResize;
exports.offWindowResize = wx.offWindowResize;
exports.onKeyboardHeightChange = wx.onKeyboardHeightChange;
exports.offKeyboardHeightChange = wx.offKeyboardHeightChange;
exports.getSelectedTextRange = wx.getSelectedTextRange;
exports.request = framework_shared_1.promisify(wx.request);
exports.downloadFile = framework_shared_1.promisify(wx.downloadFile);
exports.uploadFile = framework_shared_1.promisify(wx.uploadFile);
exports.sendSocketMessage = framework_shared_1.promisify(wx.sendSocketMessage);
exports.onSocketOpen = wx.onSocketOpen;
exports.onSocketMessage = wx.onSocketMessage;
exports.onSocketError = wx.onSocketError;
exports.onSocketClose = wx.onSocketClose;
exports.connectSocket = wx.connectSocket;
exports.closeSocket = framework_shared_1.promisify(wx.closeSocket);
exports.stopLocalServiceDiscovery = framework_shared_1.promisify(wx.stopLocalServiceDiscovery);
exports.startLocalServiceDiscovery = framework_shared_1.promisify(wx.startLocalServiceDiscovery);
exports.onLocalServiceResolveFail = wx.onLocalServiceResolveFail;
exports.onLocalServiceLost = wx.onLocalServiceLost;
exports.onLocalServiceFound = wx.onLocalServiceFound;
exports.onLocalServiceDiscoveryStop = wx.onLocalServiceDiscoveryStop;
exports.offLocalServiceResolveFail = wx.offLocalServiceResolveFail;
exports.offLocalServiceLost = wx.offLocalServiceLost;
exports.offLocalServiceFound = wx.offLocalServiceFound;
exports.offLocalServiceDiscoveryStop = wx.offLocalServiceDiscoveryStop;
exports.createUDPSocket = wx.createUDPSocket;
exports.setStorageSync = wx.setStorageSync;
exports.setStorage = framework_shared_1.promisify(wx.setStorage);
exports.removeStorageSync = wx.removeStorageSync;
exports.removeStorage = framework_shared_1.promisify(wx.removeStorage);
exports.getStorageSync = wx.getStorageSync;
exports.getStorageInfoSync = wx.getStorageInfoSync;
exports.getStorageInfo = framework_shared_1.promisify(wx.getStorageInfo);
exports.getStorage = framework_shared_1.promisify(wx.getStorage);
exports.clearStorageSync = wx.clearStorageSync;
exports.clearStorage = framework_shared_1.promisify(wx.clearStorage);
exports.createMapContext = wx.createMapContext;
exports.saveImageToPhotosAlbum = framework_shared_1.promisify(wx.saveImageToPhotosAlbum);
exports.previewImage = framework_shared_1.promisify(wx.previewImage);
exports.getImageInfo = framework_shared_1.promisify(wx.getImageInfo);
exports.compressImage = framework_shared_1.promisify(wx.compressImage);
exports.chooseMessageFile = framework_shared_1.promisify(wx.chooseMessageFile);
exports.chooseImage = framework_shared_1.promisify(wx.chooseImage);
exports.saveVideoToPhotosAlbum = framework_shared_1.promisify(wx.saveVideoToPhotosAlbum);
exports.createVideoContext = wx.createVideoContext;
exports.chooseVideo = framework_shared_1.promisify(wx.chooseVideo);
exports.chooseMedia = framework_shared_1.promisify(wx.chooseMedia);
exports.stopVoice = framework_shared_1.promisify(wx.stopVoice);
exports.setInnerAudioOption = framework_shared_1.promisify(wx.setInnerAudioOption);
exports.playVoice = framework_shared_1.promisify(wx.playVoice);
exports.pauseVoice = framework_shared_1.promisify(wx.pauseVoice);
exports.getAvailableAudioSources = framework_shared_1.promisify(wx.getAvailableAudioSources);
exports.createInnerAudioContext = wx.createInnerAudioContext;
exports.createAudioContext = wx.createAudioContext;
exports.stopBackgroundAudio = framework_shared_1.promisify(wx.stopBackgroundAudio);
exports.seekBackgroundAudio = framework_shared_1.promisify(wx.seekBackgroundAudio);
exports.playBackgroundAudio = framework_shared_1.promisify(wx.playBackgroundAudio);
exports.pauseBackgroundAudio = framework_shared_1.promisify(wx.pauseBackgroundAudio);
exports.onBackgroundAudioStop = wx.onBackgroundAudioStop;
exports.onBackgroundAudioPlay = wx.onBackgroundAudioPlay;
exports.onBackgroundAudioPause = wx.onBackgroundAudioPause;
exports.getBackgroundAudioPlayerState = framework_shared_1.promisify(wx.getBackgroundAudioPlayerState);
exports.getBackgroundAudioManager = wx.getBackgroundAudioManager;
exports.createLivePusherContext = wx.createLivePusherContext;
exports.createLivePlayerContext = wx.createLivePlayerContext;
exports.stopRecord = framework_shared_1.promisify(wx.stopRecord);
exports.startRecord = framework_shared_1.promisify(wx.startRecord);
exports.getRecorderManager = wx.getRecorderManager;
exports.createCameraContext = wx.createCameraContext;
exports.stopLocationUpdate = framework_shared_1.promisify(wx.stopLocationUpdate);
exports.startLocationUpdateBackground = framework_shared_1.promisify(wx.startLocationUpdateBackground);
exports.startLocationUpdate = framework_shared_1.promisify(wx.startLocationUpdate);
exports.openLocation = framework_shared_1.promisify(wx.openLocation);
exports.onLocationChange = wx.onLocationChange;
exports.getLocation = framework_shared_1.promisify(wx.getLocation);
exports.chooseLocation = framework_shared_1.promisify(wx.chooseLocation);
exports.updateShareMenu = framework_shared_1.promisify(wx.updateShareMenu);
exports.showShareMenu = framework_shared_1.promisify(wx.showShareMenu);
exports.hideShareMenu = framework_shared_1.promisify(wx.hideShareMenu);
exports.getShareInfo = framework_shared_1.promisify(wx.getShareInfo);
exports.createOffscreenCanvas = wx.createOffscreenCanvas;
exports.createCanvasContext = wx.createCanvasContext;
exports.canvasToTempFilePath = framework_shared_1.promisify(wx.canvasToTempFilePath);
exports.canvasPutImageData = framework_shared_1.promisify(wx.canvasPutImageData);
exports.canvasGetImageData = framework_shared_1.promisify(wx.canvasGetImageData);
exports.saveFile = framework_shared_1.promisify(wx.saveFile);
exports.removeSavedFile = framework_shared_1.promisify(wx.removeSavedFile);
exports.openDocument = framework_shared_1.promisify(wx.openDocument);
exports.getSavedFileList = framework_shared_1.promisify(wx.getSavedFileList);
exports.getSavedFileInfo = framework_shared_1.promisify(wx.getSavedFileInfo);
exports.getFileSystemManager = wx.getFileSystemManager;
exports.getFileInfo = framework_shared_1.promisify(wx.getFileInfo);
exports.login = framework_shared_1.promisify(wx.login);
exports.checkSession = framework_shared_1.promisify(wx.checkSession);
exports.navigateToMiniProgram = framework_shared_1.promisify(wx.navigateToMiniProgram);
exports.navigateBackMiniProgram = framework_shared_1.promisify(wx.navigateBackMiniProgram);
exports.getAccountInfoSync = wx.getAccountInfoSync;
exports.getUserInfo = framework_shared_1.promisify(wx.getUserInfo);
exports.reportMonitor = wx.reportMonitor;
exports.reportAnalytics = wx.reportAnalytics;
exports.requestPayment = framework_shared_1.promisify(wx.requestPayment);
exports.authorize = framework_shared_1.promisify(wx.authorize);
exports.openSetting = framework_shared_1.promisify(wx.openSetting);
exports.getSetting = framework_shared_1.promisify(wx.getSetting);
exports.chooseAddress = framework_shared_1.promisify(wx.chooseAddress);
exports.openCard = framework_shared_1.promisify(wx.openCard);
exports.addCard = framework_shared_1.promisify(wx.addCard);
exports.chooseInvoiceTitle = framework_shared_1.promisify(wx.chooseInvoiceTitle);
exports.chooseInvoice = framework_shared_1.promisify(wx.chooseInvoice);
exports.startSoterAuthentication = framework_shared_1.promisify(wx.startSoterAuthentication);
exports.checkIsSupportSoterAuthentication = framework_shared_1.promisify(wx.checkIsSupportSoterAuthentication);
exports.checkIsSoterEnrolledInDevice = framework_shared_1.promisify(wx.checkIsSoterEnrolledInDevice);
exports.getWeRunData = framework_shared_1.promisify(wx.getWeRunData);
exports.stopBeaconDiscovery = framework_shared_1.promisify(wx.stopBeaconDiscovery);
exports.startBeaconDiscovery = framework_shared_1.promisify(wx.startBeaconDiscovery);
exports.onBeaconUpdate = wx.onBeaconUpdate;
exports.offBeaconUpdate = wx.offBeaconUpdate;
exports.onBeaconServiceChange = wx.onBeaconServiceChange;
exports.offBeaconServiceChange = wx.offBeaconServiceChange;
exports.getBeacons = framework_shared_1.promisify(wx.getBeacons);
exports.stopWifi = framework_shared_1.promisify(wx.stopWifi);
exports.startWifi = framework_shared_1.promisify(wx.startWifi);
exports.setWifiList = framework_shared_1.promisify(wx.setWifiList);
exports.onWifiConnected = wx.onWifiConnected;
exports.offWifiConnected = wx.offWifiConnected;
exports.onGetWifiList = wx.onGetWifiList;
exports.offGetWifiList = wx.offGetWifiList;
exports.getWifiList = framework_shared_1.promisify(wx.getWifiList);
exports.getConnectedWifi = framework_shared_1.promisify(wx.getConnectedWifi);
exports.connectWifi = framework_shared_1.promisify(wx.connectWifi);
exports.addPhoneContact = framework_shared_1.promisify(wx.addPhoneContact);
exports.writeBLECharacteristicValue = framework_shared_1.promisify(wx.writeBLECharacteristicValue);
exports.readBLECharacteristicValue = framework_shared_1.promisify(wx.readBLECharacteristicValue);
exports.onBLEConnectionStateChange = wx.onBLEConnectionStateChange;
exports.offBLEConnectionStateChange = wx.offBLEConnectionStateChange;
exports.onBLECharacteristicValueChange = wx.onBLECharacteristicValueChange;
exports.offBLECharacteristicValueChange = wx.offBLECharacteristicValueChange;
exports.notifyBLECharacteristicValueChange = framework_shared_1.promisify(wx.notifyBLECharacteristicValueChange);
exports.getBLEDeviceServices = framework_shared_1.promisify(wx.getBLEDeviceServices);
exports.getBLEDeviceCharacteristics = framework_shared_1.promisify(wx.getBLEDeviceCharacteristics);
exports.createBLEConnection = framework_shared_1.promisify(wx.createBLEConnection);
exports.closeBLEConnection = framework_shared_1.promisify(wx.closeBLEConnection);
exports.stopBluetoothDevicesDiscovery = framework_shared_1.promisify(wx.stopBluetoothDevicesDiscovery);
exports.startBluetoothDevicesDiscovery = framework_shared_1.promisify(wx.startBluetoothDevicesDiscovery);
exports.openBluetoothAdapter = wx.openBluetoothAdapter;
exports.onBluetoothDeviceFound = wx.onBluetoothDeviceFound;
exports.offBluetoothDeviceFound = wx.offBluetoothDeviceFound;
exports.onBluetoothAdapterStateChange = wx.onBluetoothAdapterStateChange;
exports.offBluetoothAdapterStateChange = wx.offBluetoothAdapterStateChange;
exports.getConnectedBluetoothDevices = framework_shared_1.promisify(wx.getConnectedBluetoothDevices);
exports.getBluetoothDevices = framework_shared_1.promisify(wx.getBluetoothDevices);
exports.getBluetoothAdapterState = framework_shared_1.promisify(wx.getBluetoothAdapterState);
exports.closeBluetoothAdapter = framework_shared_1.promisify(wx.closeBluetoothAdapter);
exports.getBatteryInfoSync = wx.getBatteryInfoSync;
exports.getBatteryInfo = framework_shared_1.promisify(wx.getBatteryInfo);
exports.setClipboardData = framework_shared_1.promisify(wx.setClipboardData);
exports.getClipboardData = framework_shared_1.promisify(wx.getClipboardData);
exports.stopHCE = framework_shared_1.promisify(wx.stopHCE);
exports.startHCE = framework_shared_1.promisify(wx.startHCE);
exports.sendHCEMessage = framework_shared_1.promisify(wx.sendHCEMessage);
exports.onHCEMessage = wx.onHCEMessage;
exports.offHCEMessage = wx.offHCEMessage;
exports.getHCEState = framework_shared_1.promisify(wx.getHCEState);
exports.onNetworkStatusChange = wx.onNetworkStatusChange;
exports.offNetworkStatusChange = wx.offNetworkStatusChange;
exports.getNetworkType = framework_shared_1.promisify(wx.getNetworkType);
exports.setScreenBrightness = framework_shared_1.promisify(wx.setScreenBrightness);
exports.setKeepScreenOn = framework_shared_1.promisify(wx.setKeepScreenOn);
exports.onUserCaptureScreen = wx.onUserCaptureScreen;
exports.offUserCaptureScreen = wx.offUserCaptureScreen;
exports.getScreenBrightness = framework_shared_1.promisify(wx.getScreenBrightness);
exports.makePhoneCall = framework_shared_1.promisify(wx.makePhoneCall);
exports.stopAccelerometer = framework_shared_1.promisify(wx.stopAccelerometer);
exports.startAccelerometer = framework_shared_1.promisify(wx.startAccelerometer);
exports.onAccelerometerChange = wx.onAccelerometerChange;
exports.offAccelerometerChange = wx.offAccelerometerChange;
exports.stopCompass = framework_shared_1.promisify(wx.stopCompass);
exports.startCompass = framework_shared_1.promisify(wx.startCompass);
exports.onCompassChange = wx.onCompassChange;
exports.offCompassChange = wx.offCompassChange;
exports.stopDeviceMotionListening = framework_shared_1.promisify(wx.stopDeviceMotionListening);
exports.startDeviceMotionListening = framework_shared_1.promisify(wx.startDeviceMotionListening);
exports.onDeviceMotionChange = wx.onDeviceMotionChange;
exports.offDeviceMotionChange = wx.offDeviceMotionChange;
exports.stopGyroscope = framework_shared_1.promisify(wx.stopGyroscope);
exports.startGyroscope = framework_shared_1.promisify(wx.startGyroscope);
exports.onGyroscopeChange = wx.onGyroscopeChange;
exports.offGyroscopeChange = wx.offGyroscopeChange;
exports.onMemoryWarning = wx.onMemoryWarning;
exports.offMemoryWarning = wx.offMemoryWarning;
exports.scanCode = framework_shared_1.promisify(wx.scanCode);
exports.vibrateShort = framework_shared_1.promisify(wx.vibrateShort);
exports.vibrateLong = framework_shared_1.promisify(wx.vibrateLong);
exports.createWorker = wx.createWorker;
exports.getExtConfigSync = wx.getExtConfigSync;
exports.getExtConfig = framework_shared_1.promisify(wx.getExtConfig);
exports.createSelectorQuery = wx.createSelectorQuery;
exports.createIntersectionObserver = wx.createIntersectionObserver;
exports.createRewardedVideoAd = wx.createRewardedVideoAd;
exports.createInterstitialAd = wx.createInterstitialAd;
exports.cloud = wx.cloud;
exports.requestSubscribeMessage = framework_shared_1.promisify(wx.requestSubscribeMessage);
exports.hideHomeButton = framework_shared_1.promisify(wx.hideHomeButton);
exports.getRealtimeLogManager = wx.getRealtimeLogManager;
exports.hideKeyboard = framework_shared_1.promisify(wx.hideKeyboard);
